<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Call haplotype specific copy number in single cell datasets — callHaplotypeSpecificCN • signals</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Call haplotype specific copy number in single cell datasets — callHaplotypeSpecificCN"><meta property="og:description" content="Call haplotype specific copy number in single cell datasets"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Call haplotype specific copy number in single cell datasets</h1>
    
    <div class="hidden name"><code>callHaplotypeSpecificCN.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Call haplotype specific copy number in single cell datasets</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">callHaplotypeSpecificCN</span><span class="op">(</span>
  <span class="va">CNbins</span>,
  <span class="va">haplotypes</span>,
  eps <span class="op">=</span> <span class="fl">0.000000000001</span>,
  loherror <span class="op">=</span> <span class="fl">0.02</span>,
  maxCN <span class="op">=</span> <span class="cn">NULL</span>,
  selftransitionprob <span class="op">=</span> <span class="fl">0.95</span>,
  progressbar <span class="op">=</span> <span class="cn">TRUE</span>,
  ncores <span class="op">=</span> <span class="fl">1</span>,
  phasebyarm <span class="op">=</span> <span class="cn">FALSE</span>,
  minfrac <span class="op">=</span> <span class="fl">0.7</span>,
  likelihood <span class="op">=</span> <span class="st">"binomial"</span>,
  minbins <span class="op">=</span> <span class="fl">100</span>,
  minbinschr <span class="op">=</span> <span class="fl">10</span>,
  phased_haplotypes <span class="op">=</span> <span class="cn">NULL</span>,
  clustering_method <span class="op">=</span> <span class="st">"copy"</span>,
  maxloherror <span class="op">=</span> <span class="fl">0.035</span>,
  mincells <span class="op">=</span> <span class="fl">10</span>,
  overwritemincells <span class="op">=</span> <span class="cn">NULL</span>,
  cluster_per_chr <span class="op">=</span> <span class="cn">TRUE</span>,
  viterbiver <span class="op">=</span> <span class="st">"cpp"</span>,
  filterhaplotypes <span class="op">=</span> <span class="fl">0.1</span>,
  firstpassfiltering <span class="op">=</span> <span class="cn">TRUE</span>,
  smoothsingletons <span class="op">=</span> <span class="cn">TRUE</span>,
  fillmissing <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>CNbins</dt>
<dd><p>single cell copy number dataframe with the following columns: <code>cell_id</code>, <code>chr</code>, <code>start</code>, <code>end</code>, <code>state</code>, <code>copy</code></p></dd>
<dt>haplotypes</dt>
<dd><p>single cell haplotypes dataframe with the following columns: <code>cell_id</code>, <code>chr</code>, <code>start</code>, <code>end</code>, <code>hap_label</code>, <code>allele1</code>, <code>allele0</code>, <code>totalcounts</code></p></dd>
<dt>eps</dt>
<dd><p>default 1e-12</p></dd>
<dt>loherror</dt>
<dd><p>LOH error rate for initial assignment, this is inferred directly from the data in the second pass, default = 0.02</p></dd>
<dt>maxCN</dt>
<dd><p>maximum copy number to infer allele specific states, default=NULL which will use the maximum state from CNbins</p></dd>
<dt>selftransitionprob</dt>
<dd><p>probability to stay in the same state in the HMM, default = 0.95, set to 0.0 for an IID model</p></dd>
<dt>progressbar</dt>
<dd><p>Boolean to display progressbar or not, default = TRUE, will only show if ncores == 1</p></dd>
<dt>ncores</dt>
<dd><p>Number of cores to use, default = 1</p></dd>
<dt>minfrac</dt>
<dd><p>Minimum proportion of haplotypes to retain when clustering + phasing, default = 0.8</p></dd>
<dt>likelihood</dt>
<dd><p>Likelihood model for HMM, default is <code>binomial</code>, other option is <code>betabinomial</code> or use <code>auto</code> and the algorithm will choose the likelihood that best fits the data.</p></dd>
<dt>minbins</dt>
<dd><p>Minimum number of bins containing both haplotype counts and copy number data for a cell to be included</p></dd>
<dt>minbinschr</dt>
<dd><p>Minimum number of bins containing both haplotype counts and copy number data per chromosome for a cell to be included</p></dd>
<dt>phased_haplotypes</dt>
<dd><p>Use this if you want to manually define the haplotypes phasing if for example the default heuristics used by schnapps does not return a good fit.</p></dd>
<dt>clustering_method</dt>
<dd><p>Method to use to cluster cells for haplotype phasing, default is <code>copy</code>, other option is <code>breakpoints</code></p></dd>
<dt>maxloherror</dt>
<dd><p>Maximum value for LOH error rate</p></dd>
<dt>mincells</dt>
<dd><p>Minimum cluster size used for phasing</p></dd>
<dt>overwritemincells</dt>
<dd><p>Overwrite the the number of cells to use for clustering/phasing</p></dd>
<dt>cluster_per_chr</dt>
<dd><p>Whether to cluster per chromosome to rephase alleles or not</p></dd>
<dt>filterhaplotypes</dt>
<dd><p>filter out haplotypes present in less than X fraction, default is 0.1</p></dd>
<dt>firstpassfiltering</dt>
<dd><p>Filter out cells with large discrepancy after first pass state assignment</p></dd>
<dt>smoothsingletons</dt>
<dd><p>Remove singleton bins by smoothing over based on states in adjacent bins</p></dd>
<dt>fillmissing</dt>
<dd><p>For bins with missing counts fill in values based on neighbouring bins</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>allele specific copy number object which includes dataframe similar to input with additional columns which include</p><ul><li><p><code>Maj</code> (Major allele copy number)</p></li>
<li><p><code>Min</code> (Minor allele copy number)</p></li>
<li><p><code>state_AS_phased</code> (phased state of the form Maj|Min )</p></li>
<li><p><code>state_AS</code> (mirrored state of the form Maj|Min)</p></li>
<li><p><code>LOH</code> (is bin LOH or not)</p></li>
<li><p><code>state_phase</code> (state describing which is the dominant allele and whether it is LOH or not)</p></li>
<li><p><code>state_BAF</code> (binned discretized BAF value calculated as Min / (Maj + Min))</p></li>
</ul></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">sim_data</span> <span class="op">&lt;-</span> <span class="fu">simulate_data_cohort</span><span class="op">(</span></span>
<span class="r-in">  clone_num <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">20</span><span class="op">)</span>,</span>
<span class="r-in">  clonal_events <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span class="r-in">    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"1"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0</span><span class="op">)</span>, <span class="st">"5"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>,</span>
<span class="r-in">    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"2"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">6</span>, <span class="fl">3</span><span class="op">)</span>, <span class="st">"3"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in">  <span class="op">)</span>,</span>
<span class="r-in">  loherror <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span class="r-in">  coverage <span class="op">=</span> <span class="fl">100</span></span>
<span class="r-in"><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining, by = c("chr", "start", "end", "hap_label")</span>
<span class="r-in"></span>
<span class="r-in"><span class="va">results</span> <span class="op">&lt;-</span> <span class="fu">callHaplotypeSpecificCN</span><span class="op">(</span><span class="va">sim_data</span><span class="op">$</span><span class="va">CNbins</span>, <span class="va">sim_data</span><span class="op">$</span><span class="va">haplotypes</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering out haplotypes present &lt; 10% of cells...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Fraction of haplotypes retained after filtering = 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding overlapping cell IDs between CN data and haplotype data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells in both CN and haplotypes: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in CN data: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in haplotype data: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining bins and haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Phase haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Phasing based on distribution across all cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Join phased haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reorder haplotypes based on phase...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells after removing cells with &lt; 100 bins: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing 0 cells for phasing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using 10 cells for clustering...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 2 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster A: 24</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster B: 16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 10</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 11</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 12</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 13</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 14</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 15</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 16</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 17</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 18</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 19</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 20</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 21</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 22</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 2 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster A: 23</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster B: 17</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 2 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster A: 22</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster B: 18</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 7</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 8</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome 9</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome X</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering chromosome Y</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating CN matrix...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating UMAP dimensionality reduction...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Clustering cells using hdbscan...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Identified 1 clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Distribution of clusters:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   Cluster 0: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding overlapping cell IDs between CN data and haplotype data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells in both CN and haplotypes: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in CN data: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in haplotype data: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining bins and haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Phase haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Join phased haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reorder haplotypes based on phase...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells after removing cells with &lt; 100 bins: 40</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total fraction of haplotype counts retained: 1 (Total counts (Millions): 0.94550355, after filtering: 0.94550355, after phasing:0.94550355)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Average distance from median to expected BAF = 0.0053</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Average distance from median to expected BAF = 0.0053</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

