<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Call allele specific copy number in single cell datasets — callAlleleSpecificCN • signals</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Call allele specific copy number in single cell datasets — callAlleleSpecificCN"><meta property="og:description" content="Call allele specific copy number in single cell datasets"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/heatmap_visualization.html">Advanced Heatmap Visualization</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Call allele specific copy number in single cell datasets</h1>

    <div class="hidden name"><code>callAlleleSpecificCN.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Call allele specific copy number in single cell datasets</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">callAlleleSpecificCN</span><span class="op">(</span></span>
<span>  <span class="va">CNbins</span>,</span>
<span>  <span class="va">haplotypes</span>,</span>
<span>  eps <span class="op">=</span> <span class="fl">1e-12</span>,</span>
<span>  loherror <span class="op">=</span> <span class="fl">0.02</span>,</span>
<span>  maxCN <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  selftransitionprob <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  progressbar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ncores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  likelihood <span class="op">=</span> <span class="st">"binomial"</span>,</span>
<span>  minbins <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  minbinschr <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxloherror <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  filterhaplotypes <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  fillmissing <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-cnbins">CNbins<a class="anchor" aria-label="anchor" href="#arg-cnbins"></a></dt>
<dd><p>single cell copy number dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `state`, `copy`</p></dd>


<dt id="arg-haplotypes">haplotypes<a class="anchor" aria-label="anchor" href="#arg-haplotypes"></a></dt>
<dd><p>single cell haplotypes dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `hap_label`, `allele1`, `allele0`, `totalcounts`</p></dd>


<dt id="arg-eps">eps<a class="anchor" aria-label="anchor" href="#arg-eps"></a></dt>
<dd><p>default 1e-12</p></dd>


<dt id="arg-loherror">loherror<a class="anchor" aria-label="anchor" href="#arg-loherror"></a></dt>
<dd><p>LOH error rate for initial assignment, this is inferred directly from the data in the second pass, default = 0.02</p></dd>


<dt id="arg-maxcn">maxCN<a class="anchor" aria-label="anchor" href="#arg-maxcn"></a></dt>
<dd><p>maximum copy number to infer allele specific states, default=NULL which will use the maximum state from CNbins</p></dd>


<dt id="arg-selftransitionprob">selftransitionprob<a class="anchor" aria-label="anchor" href="#arg-selftransitionprob"></a></dt>
<dd><p>probability to stay in the same state in the HMM, default = 0.999, set to 0.0 for an IID model</p></dd>


<dt id="arg-progressbar">progressbar<a class="anchor" aria-label="anchor" href="#arg-progressbar"></a></dt>
<dd><p>Boolean to display progressbar or not, default = TRUE, will only show if ncores == 1</p></dd>


<dt id="arg-ncores">ncores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>Number of cores to use, default = 1</p></dd>


<dt id="arg-likelihood">likelihood<a class="anchor" aria-label="anchor" href="#arg-likelihood"></a></dt>
<dd><p>Likelihood model for HMM, default is `binomial`, other option is `betabinomial` or use `auto` and the algorithm will choose the likelihood that best fits the data.</p></dd>


<dt id="arg-minbins">minbins<a class="anchor" aria-label="anchor" href="#arg-minbins"></a></dt>
<dd><p>Minimum number of bins containing both haplotype counts and copy number data for a cell to be included</p></dd>


<dt id="arg-minbinschr">minbinschr<a class="anchor" aria-label="anchor" href="#arg-minbinschr"></a></dt>
<dd><p>Minimum number of bins containing both haplotype counts and copy number data per chromosome for a cell to be included</p></dd>


<dt id="arg-maxloherror">maxloherror<a class="anchor" aria-label="anchor" href="#arg-maxloherror"></a></dt>
<dd><p>Maximum value for LOH error rate</p></dd>


<dt id="arg-filterhaplotypes">filterhaplotypes<a class="anchor" aria-label="anchor" href="#arg-filterhaplotypes"></a></dt>
<dd><p>filter out haplotypes present in less than X fraction, default is 0.1</p></dd>


<dt id="arg-fillmissing">fillmissing<a class="anchor" aria-label="anchor" href="#arg-fillmissing"></a></dt>
<dd><p>For bins with missing counts fill in values based on neighbouring bins</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>allele specific copy number object which includes dataframe similar to input with additional columns which include</p>
<p>* `A` A allele copy number
* `B` B allele copy number
* `state_AS_phased` A|B
* `state_min` Minor allele copy number
* `LOH` =LOH if bin is LOH, NO otherwise
* `state_phase` Discretized haplotype specific states
* `phase` Whether the A allele or B allele is dominant
* `alleleA` Counts for the A allele
* `alleleB` Counts for the B allele
* `totalcounts` Total number of counts
* `BAF` B-allele frequency (alleleB / totalcounts)</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>In the allele specific copy number inference A is always &gt; B and state_AS_phased == state_AS</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

