<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Combine copy number bins with haplotype data — combineBAFCN • signals</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Combine copy number bins with haplotype data — combineBAFCN"><meta property="og:description" content="Joins copy number bin data with phased haplotype counts to produce a combined
data.frame with copy number states and B-allele frequency (BAF) for each bin."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.12.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Combine copy number bins with haplotype data</h1>

    <div class="hidden name"><code>combineBAFCN.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Joins copy number bin data with phased haplotype counts to produce a combined
data.frame with copy number states and B-allele frequency (BAF) for each bin.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">combineBAFCN</span><span class="op">(</span></span>
<span>  <span class="va">haplotypes</span>,</span>
<span>  <span class="va">CNbins</span>,</span>
<span>  filtern <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  phased_haplotypes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minbins <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  minbinschr <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  phasing_method <span class="op">=</span> <span class="st">"distribution"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-haplotypes">haplotypes<a class="anchor" aria-label="anchor" href="#arg-haplotypes"></a></dt>
<dd><p>A data.frame with haplotype allele counts. Required columns:
`cell_id`, `chr`, `start`, `end`, `hap_label`, `allele_id`, `readcount` (raw) or
`allele1`, `allele0`, `totalcounts` (formatted).</p></dd>


<dt id="arg-cnbins">CNbins<a class="anchor" aria-label="anchor" href="#arg-cnbins"></a></dt>
<dd><p>A data.frame with copy number bin data. Required columns:
`cell_id`, `chr`, `start`, `end`, `state`, `copy`.</p></dd>


<dt id="arg-filtern">filtern<a class="anchor" aria-label="anchor" href="#arg-filtern"></a></dt>
<dd><p>Minimum total read count per bin to include. Default 0.</p></dd>


<dt id="arg-phased-haplotypes">phased_haplotypes<a class="anchor" aria-label="anchor" href="#arg-phased-haplotypes"></a></dt>
<dd><p>Optional pre-computed phased haplotypes from
`computehaplotypecounts()`. If NULL, phasing is performed automatically.</p></dd>


<dt id="arg-minbins">minbins<a class="anchor" aria-label="anchor" href="#arg-minbins"></a></dt>
<dd><p>Minimum number of bins per cell to include. Default 100.</p></dd>


<dt id="arg-minbinschr">minbinschr<a class="anchor" aria-label="anchor" href="#arg-minbinschr"></a></dt>
<dd><p>Minimum number of bins per chromosome per cell. Default 10.</p></dd>


<dt id="arg-phasing-method">phasing_method<a class="anchor" aria-label="anchor" href="#arg-phasing-method"></a></dt>
<dd><p>Method for phasing haplotypes. One of "distribution" (default)
or use top N imbalanced cells.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to `format_haplotypes()`.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A data.frame with columns from CNbins plus:
  * `alleleA`: Read counts for A allele
  * `alleleB`: Read counts for B allele
  * `totalcounts`: Total read counts
  * `BAF`: B-allele frequency (alleleB / totalcounts)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">haplotypes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Format haplotypes first</span></span></span>
<span class="r-in"><span><span class="va">haps_formatted</span> <span class="op">&lt;-</span> <span class="fu"><a href="format_haplotypes_dlp.html">format_haplotypes_dlp</a></span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of distinct bins in copy number data: 4375</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of distinct bins in haplotype data: 5728</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of distinct bins in formatted haplotype data: 4360</span>
<span class="r-in"><span><span class="co"># Combine with CNbins</span></span></span>
<span class="r-in"><span><span class="va">cnbaf</span> <span class="op">&lt;-</span> <span class="fu">combineBAFCN</span><span class="op">(</span><span class="va">haps_formatted</span>, <span class="va">CNbins</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Finding overlapping cell IDs between CN data and haplotype data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells in both CN and haplotypes: 250</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in CN data: 250</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Number of cells in haplotype data: 250</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining bins and haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Phase haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Phasing based on distribution across all cells</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Join phased haplotypes...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reorder haplotypes based on phase...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Total number of cells after removing cells with &lt; 100 bins: 250</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

