<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot a single cell allele specific copy number profile — plotCNprofileBAF • schnapps</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Plot a single cell allele specific copy number profile — plotCNprofileBAF" />
<meta property="og:description" content="Plot a single cell allele specific copy number profile" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">schnapps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for schnapps</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot a single cell allele specific copy number profile</h1>
    
    <div class="hidden name"><code>plotCNprofileBAF.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a single cell allele specific copy number profile</p>
    </div>

    <pre class="usage"><span class='fu'>plotCNprofileBAF</span><span class='op'>(</span>
  <span class='va'>cn</span>,
  cellid <span class='op'>=</span> <span class='cn'>NULL</span>,
  chrfilt <span class='op'>=</span> <span class='cn'>NULL</span>,
  pointsize <span class='op'>=</span> <span class='fl'>1</span>,
  alphaval <span class='op'>=</span> <span class='fl'>0.6</span>,
  maxCN <span class='op'>=</span> <span class='fl'>10</span>,
  cellidx <span class='op'>=</span> <span class='fl'>1</span>,
  BAFcol <span class='op'>=</span> <span class='st'>"state_phase"</span>,
  statecol <span class='op'>=</span> <span class='st'>"state"</span>,
  returnlist <span class='op'>=</span> <span class='cn'>FALSE</span>,
  raster <span class='op'>=</span> <span class='cn'>FALSE</span>,
  y_axis_trans <span class='op'>=</span> <span class='st'>"identity"</span>,
  xaxis_order <span class='op'>=</span> <span class='st'>"genome_position"</span>,
  legend.position <span class='op'>=</span> <span class='st'>"bottom"</span>,
  genes <span class='op'>=</span> <span class='cn'>NULL</span>,
  annotateregions <span class='op'>=</span> <span class='cn'>NULL</span>,
  homolog <span class='op'>=</span> <span class='cn'>FALSE</span>,
  SV <span class='op'>=</span> <span class='cn'>NULL</span>,
  adj <span class='op'>=</span> <span class='fl'>0.03</span>,
  svalpha <span class='op'>=</span> <span class='fl'>0.5</span>,
  svwidth <span class='op'>=</span> <span class='fl'>1</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cn</th>
      <td><p>Single cell allele specific copy number dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `state`, `copy` or a hscn object.</p></td>
    </tr>
    <tr>
      <th>cellid</th>
      <td><p>Which cell to plot, if no cell is specific will plot the first cell in the dataframe</p></td>
    </tr>
    <tr>
      <th>chrfilt</th>
      <td><p>Vector of chromosomes to plot, if NULL (default) will plot all chromosomes</p></td>
    </tr>
    <tr>
      <th>pointsize</th>
      <td><p>The point size in the plot</p></td>
    </tr>
    <tr>
      <th>alphaval</th>
      <td><p>Alpha value of points</p></td>
    </tr>
    <tr>
      <th>maxCN</th>
      <td><p>The maximum on the y axis, if any points are above this value they will be winsorized rather than removed</p></td>
    </tr>
    <tr>
      <th>cellidx</th>
      <td><p>idx of cell to plot if cellid = NULL</p></td>
    </tr>
    <tr>
      <th>statecol</th>
      <td><p>The colour mapping, default is to map colours to the `state` column</p></td>
    </tr>
    <tr>
      <th>returnlist</th>
      <td><p>Return a list rather than the ggplot object</p></td>
    </tr>
    <tr>
      <th>raster</th>
      <td><p>use ggrastr or not, default = FALSE</p></td>
    </tr>
    <tr>
      <th>y_axis_trans</th>
      <td><p>What transformation to use on the y-axis, default is identity, the other option is "squashy" which uses a tanh transformation</p></td>
    </tr>
    <tr>
      <th>xaxis_order</th>
      <td><p>Default is "genome_position"</p></td>
    </tr>
    <tr>
      <th>legend.position</th>
      <td><p>Where to place the legend, default is "bottom"</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>vector of genes to annotate, will add a dashed vertical line and label</p></td>
    </tr>
    <tr>
      <th>annotateregions</th>
      <td><p>Dataframe with chr start and end positions to annotate, will draw a dashed vertical line at this position</p></td>
    </tr>
    <tr>
      <th>homolog</th>
      <td><p>Rather than plot the BAF and CN seperately this will plot the 2 homologs on the same track</p></td>
    </tr>
    <tr>
      <th>SV</th>
      <td><p>Default is NULL. If a dataframe with structural variant position is passed it will add a track on the top showin rearrangement links</p></td>
    </tr>
    <tr>
      <th>svalpha</th>
      <td><p>the alpha scaling of the SV lines, default = 0.5</p></td>
    </tr>
    <tr>
      <th>svwidth</th>
      <td><p>width of rearrangement connections</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>ggplot2 plot</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='r-in'></span>
<span class='r-in'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"haplotypes"</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='st'>"CNbins"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>haplotypes</span> <span class='op'>&lt;-</span> <span class='fu'>format_haplotypes_dlp</span><span class='op'>(</span><span class='va'>haplotypes</span>, <span class='va'>CNbins</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>hscn</span> <span class='op'>&lt;-</span> <span class='fu'><a href='callHaplotypeSpecificCN.html'>callHaplotypeSpecificCN</a></span><span class='op'>(</span><span class='va'>CNbins</span>, <span class='va'>haplotypes</span>, likelihood <span class='op'>=</span> <span class='st'>"binomial"</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>plotCNprofileBAF</span><span class='op'>(</span><span class='va'>hscn</span>, genes <span class='op'>=</span> <span class='st'>"MYC"</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>plotCNprofileBAF</span><span class='op'>(</span><span class='va'>hscn</span>, homolog <span class='op'>=</span> <span class='cn'>TRUE</span>, chrfilt <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"1"</span>, <span class='st'>"8"</span><span class='op'>)</span><span class='op'>)</span></span>
<span class='r-in'><span class='op'>}</span></span>
<span class='r-in'></span>
</pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


