<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Heatmap plot — plotHeatmap • signals</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap plot — plotHeatmap"><meta property="og:description" content="Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.11.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heatmap plot</h1>

    <div class="hidden name"><code>plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotHeatmap</span><span class="op">(</span></span>
<span>  <span class="va">cn</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalize_ploidy <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  normalize_tree <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  branch_length <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  spacer_cols <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  plottree <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plotcol <span class="op">=</span> <span class="st">"state"</span>,</span>
<span>  reorderclusters <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pctcells <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  library_mapping <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clone_pal <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_label_idx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  fillna <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  frequencycutoff <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  frequency_bar_width <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  maxf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plotfrequency <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  frequency_height <span class="op">=</span> <span class="fl">1.4</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_library_label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_clone_label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_clone_text <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  widenarm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  umapmetric <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  chrlabels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  labeladjust <span class="op">=</span> <span class="op">-</span><span class="fl">5</span>,</span>
<span>  SV <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  nticks <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  Mb <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  fillgenome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  annotation_height <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annofontsize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  na_col <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  linkheight <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>  newlegendname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  str_to_remove <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxCNcol <span class="op">=</span> <span class="fl">11</span>,</span>
<span>  anno_width <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>  rasterquality <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  tree_width <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-cn">cn<a class="anchor" aria-label="anchor" href="#arg-cn"></a></dt>
<dd><p>Either a hscn object or a single cell allele specific copy number dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `state`, `copy`</p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>Tree in newick format to plot alongside the heatmap, default = NULL</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>data.frame assigning cells to clusters, needs the following columns `cell_id`, `clone_id` default = NULL</p></dd>


<dt id="arg-normalize-ploidy">normalize_ploidy<a class="anchor" aria-label="anchor" href="#arg-normalize-ploidy"></a></dt>
<dd><p>Normalize ploidy of all cells to 2</p></dd>


<dt id="arg-normalize-tree">normalize_tree<a class="anchor" aria-label="anchor" href="#arg-normalize-tree"></a></dt>
<dd><p>default = FALSE</p></dd>


<dt id="arg-branch-length">branch_length<a class="anchor" aria-label="anchor" href="#arg-branch-length"></a></dt>
<dd><p>scales branch lengths to this size, default = 2</p></dd>


<dt id="arg-spacer-cols">spacer_cols<a class="anchor" aria-label="anchor" href="#arg-spacer-cols"></a></dt>
<dd><p>number of empty columns between chromosomes, default = 20</p></dd>


<dt id="arg-plottree">plottree<a class="anchor" aria-label="anchor" href="#arg-plottree"></a></dt>
<dd><p>Binary value of whether to plot tree or not, default = TRUE</p></dd>


<dt id="arg-plotcol">plotcol<a class="anchor" aria-label="anchor" href="#arg-plotcol"></a></dt>
<dd><p>Which column to colour the heatmap by, should be one of "state", "state_BAF", "state_phase", "state_AS", "state_min", "copy", "BAF", "A", "B"</p></dd>


<dt id="arg-reorderclusters">reorderclusters<a class="anchor" aria-label="anchor" href="#arg-reorderclusters"></a></dt>
<dd><p>Reorder the cells according to cluster if no tree is specified</p></dd>


<dt id="arg-pctcells">pctcells<a class="anchor" aria-label="anchor" href="#arg-pctcells"></a></dt>
<dd><p>Minimum size of cluster in terms of perecentage of cells in umap clustering</p></dd>


<dt id="arg-library-mapping">library_mapping<a class="anchor" aria-label="anchor" href="#arg-library-mapping"></a></dt>
<dd><p>Named vector mapping library names to labels for legend</p></dd>


<dt id="arg-clone-pal">clone_pal<a class="anchor" aria-label="anchor" href="#arg-clone-pal"></a></dt>
<dd><p>pallette to colour clusters by</p></dd>


<dt id="arg-sample-label-idx">sample_label_idx<a class="anchor" aria-label="anchor" href="#arg-sample-label-idx"></a></dt>
<dd><p>default = 1</p></dd>


<dt id="arg-fillna">fillna<a class="anchor" aria-label="anchor" href="#arg-fillna"></a></dt>
<dd><p>Smooth over NA values, default = TRUE</p></dd>


<dt id="arg-frequencycutoff">frequencycutoff<a class="anchor" aria-label="anchor" href="#arg-frequencycutoff"></a></dt>
<dd><p>default = 2</p></dd>


<dt id="arg-frequency-bar-width">frequency_bar_width<a class="anchor" aria-label="anchor" href="#arg-frequency-bar-width"></a></dt>
<dd><p>Width of bars for frequency track, default = 0.5</p></dd>


<dt id="arg-maxf">maxf<a class="anchor" aria-label="anchor" href="#arg-maxf"></a></dt>
<dd><p>Max frequency when plotting the frequency track, default = NULL infers this from the data</p></dd>


<dt id="arg-plotfrequency">plotfrequency<a class="anchor" aria-label="anchor" href="#arg-plotfrequency"></a></dt>
<dd><p>Plot the frequency track of gains and losses across the genome</p></dd>


<dt id="arg-frequency-height">frequency_height<a class="anchor" aria-label="anchor" href="#arg-frequency-height"></a></dt>
<dd><p>height of the frequency track if using, default = 1.4</p></dd>


<dt id="arg-show-legend">show_legend<a class="anchor" aria-label="anchor" href="#arg-show-legend"></a></dt>
<dd><p>plot legend or not, boolean</p></dd>


<dt id="arg-show-library-label">show_library_label<a class="anchor" aria-label="anchor" href="#arg-show-library-label"></a></dt>
<dd><p>show library label or not, boolean</p></dd>


<dt id="arg-show-clone-label">show_clone_label<a class="anchor" aria-label="anchor" href="#arg-show-clone-label"></a></dt>
<dd><p>show clone label or not, boolean</p></dd>


<dt id="arg-show-clone-text">show_clone_text<a class="anchor" aria-label="anchor" href="#arg-show-clone-text"></a></dt>
<dd><p>Show small inset labels next to clone/cluster annotation</p></dd>


<dt id="arg-widenarm">widenarm<a class="anchor" aria-label="anchor" href="#arg-widenarm"></a></dt>
<dd><p>Widen the copy number data table to include all bins</p></dd>


<dt id="arg-umapmetric">umapmetric<a class="anchor" aria-label="anchor" href="#arg-umapmetric"></a></dt>
<dd><p>metric to use in umap dimensionality reduction if no clusters are specified</p></dd>


<dt id="arg-chrlabels">chrlabels<a class="anchor" aria-label="anchor" href="#arg-chrlabels"></a></dt>
<dd><p>include chromosome labels or not, boolean</p></dd>


<dt id="arg-labeladjust">labeladjust<a class="anchor" aria-label="anchor" href="#arg-labeladjust"></a></dt>
<dd></dd>


<dt id="arg-sv">SV<a class="anchor" aria-label="anchor" href="#arg-sv"></a></dt>
<dd><p>sv data frame</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>seed for UMAP</p></dd>


<dt id="arg-nticks">nticks<a class="anchor" aria-label="anchor" href="#arg-nticks"></a></dt>
<dd><p>number of ticks in x-axis label when plotting a single chromosome</p></dd>


<dt id="arg-mb">Mb<a class="anchor" aria-label="anchor" href="#arg-mb"></a></dt>
<dd><p>Use Mb ticks when plotting single chromosome</p></dd>


<dt id="arg-fillgenome">fillgenome<a class="anchor" aria-label="anchor" href="#arg-fillgenome"></a></dt>
<dd><p>fill in any missing bins and add NA to centromeric regions</p></dd>


<dt id="arg-annotation-height">annotation_height<a class="anchor" aria-label="anchor" href="#arg-annotation-height"></a></dt>
<dd><p>Height of the annotations</p></dd>


<dt id="arg-annofontsize">annofontsize<a class="anchor" aria-label="anchor" href="#arg-annofontsize"></a></dt>
<dd><p>Font size to use for annotations, default = 10</p></dd>


<dt id="arg-na-col">na_col<a class="anchor" aria-label="anchor" href="#arg-na-col"></a></dt>
<dd><p>colour of NA values</p></dd>


<dt id="arg-linkheight">linkheight<a class="anchor" aria-label="anchor" href="#arg-linkheight"></a></dt>
<dd><p>height of x-axis ticks</p></dd>


<dt id="arg-newlegendname">newlegendname<a class="anchor" aria-label="anchor" href="#arg-newlegendname"></a></dt>
<dd><p>overwrite default legend name</p></dd>


<dt id="arg-str-to-remove">str_to_remove<a class="anchor" aria-label="anchor" href="#arg-str-to-remove"></a></dt>
<dd><p>string to remove from cell_id's when plotting labels</p></dd>


<dt id="arg-maxcncol">maxCNcol<a class="anchor" aria-label="anchor" href="#arg-maxcncol"></a></dt>
<dd><p>max value for color scale when plotting raw data</p></dd>


<dt id="arg-anno-width">anno_width<a class="anchor" aria-label="anchor" href="#arg-anno-width"></a></dt>
<dd><p>width of left annotations</p></dd>


<dt id="arg-rasterquality">rasterquality<a class="anchor" aria-label="anchor" href="#arg-rasterquality"></a></dt>
<dd><p>default = 15</p></dd>


<dt id="arg-tree-width">tree_width<a class="anchor" aria-label="anchor" href="#arg-tree-width"></a></dt>
<dd><p>Width of phylogenetic tree, default = 4</p>
<p>If clusters are set to NULL then the function will compute clusters using UMAP and HDBSCAN.</p></dd>

</dl></div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"haplotypes"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CNbins"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span>, likelihood <span class="op">=</span> <span class="st">"binomial"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">plotHeatmap</span><span class="op">(</span><span class="va">hscn</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

