<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot a single cell copy number profile — plotCNprofile • signals</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot a single cell copy number profile — plotCNprofile"><meta property="og:description" content="Plot a single cell copy number profile"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.13.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot a single cell copy number profile</h1>

    <div class="hidden name"><code>plotCNprofile.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a single cell copy number profile</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotCNprofile</span><span class="op">(</span></span>
<span>  <span class="va">CNbins</span>,</span>
<span>  cellid <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrfilt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pointsize <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  alphaval <span class="op">=</span> <span class="fl">0.6</span>,</span>
<span>  maxCN <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  cellidx <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  statecol <span class="op">=</span> <span class="st">"state"</span>,</span>
<span>  returnlist <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  raster <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"hg19"</span>,</span>
<span>  y_axis_trans <span class="op">=</span> <span class="st">"identity"</span>,</span>
<span>  xaxis_order <span class="op">=</span> <span class="st">"genome_position"</span>,</span>
<span>  legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  annotateregions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  annotateregions_linetype <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  SV <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  SVcol <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  svalpha <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  svwidth <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  adj <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tickwidth <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  chrstart <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  shape <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  positionticks <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ideogram <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  overwrite_color <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sv_style <span class="op">=</span> <span class="st">"curves"</span>,</span>
<span>  sv_arc_alpha <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  sv_arc_height <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  show_sv_read_axis <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sv_read_axis_scale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_sv_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-cnbins">CNbins<a class="anchor" aria-label="anchor" href="#arg-cnbins"></a></dt>
<dd><p>Single cell copy number dataframe with the following columns: <code>cell_id</code>, <code>chr</code>, <code>start</code>, <code>end</code>, <code>state</code>, <code>copy</code></p></dd>


<dt id="arg-cellid">cellid<a class="anchor" aria-label="anchor" href="#arg-cellid"></a></dt>
<dd><p>Which cell to plot, if no cell is specific will plot the first cell in the dataframe</p></dd>


<dt id="arg-chrfilt">chrfilt<a class="anchor" aria-label="anchor" href="#arg-chrfilt"></a></dt>
<dd><p>Vector of chromosomes to plot, if NULL (default) will plot all chromosomes</p></dd>


<dt id="arg-pointsize">pointsize<a class="anchor" aria-label="anchor" href="#arg-pointsize"></a></dt>
<dd><p>The point size in the plot</p></dd>


<dt id="arg-alphaval">alphaval<a class="anchor" aria-label="anchor" href="#arg-alphaval"></a></dt>
<dd><p>Alpha value of points</p></dd>


<dt id="arg-maxcn">maxCN<a class="anchor" aria-label="anchor" href="#arg-maxcn"></a></dt>
<dd><p>The maximum on the y axis, if any points are above this value they will be winsorized rather than removed</p></dd>


<dt id="arg-cellidx">cellidx<a class="anchor" aria-label="anchor" href="#arg-cellidx"></a></dt>
<dd><p>idx of cell to plot if cellid = NULL</p></dd>


<dt id="arg-statecol">statecol<a class="anchor" aria-label="anchor" href="#arg-statecol"></a></dt>
<dd><p>The colour mapping, default is to map colours to the <code>state</code> column</p></dd>


<dt id="arg-returnlist">returnlist<a class="anchor" aria-label="anchor" href="#arg-returnlist"></a></dt>
<dd><p>Return a list rather than the ggplot object</p></dd>


<dt id="arg-raster">raster<a class="anchor" aria-label="anchor" href="#arg-raster"></a></dt>
<dd><p>use ggrastr or not, default = FALSE</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>genome to use, default = "hg19" (only used for ideogram)</p></dd>


<dt id="arg-y-axis-trans">y_axis_trans<a class="anchor" aria-label="anchor" href="#arg-y-axis-trans"></a></dt>
<dd><p>What transformation to use on the y-axis, default is identity, the other option is "squashy" which uses a tanh transformation</p></dd>


<dt id="arg-xaxis-order">xaxis_order<a class="anchor" aria-label="anchor" href="#arg-xaxis-order"></a></dt>
<dd><p>Default is "genome_position"</p></dd>


<dt id="arg-legend-position">legend.position<a class="anchor" aria-label="anchor" href="#arg-legend-position"></a></dt>
<dd><p>Where to place the legend, default is "bottom"</p></dd>


<dt id="arg-annotateregions">annotateregions<a class="anchor" aria-label="anchor" href="#arg-annotateregions"></a></dt>
<dd><p>Dataframe with chr start and end positions to annotate, will draw a dashed vertical line at this position</p></dd>


<dt id="arg-annotateregions-linetype">annotateregions_linetype<a class="anchor" aria-label="anchor" href="#arg-annotateregions-linetype"></a></dt>
<dd><p>linetype for region annotation, default = 2 (dashed)</p></dd>


<dt id="arg-sv">SV<a class="anchor" aria-label="anchor" href="#arg-sv"></a></dt>
<dd><p>Default is NULL. If a dataframe with structural variant position is passed it will add rearrangement links between bins. For lines_and_arcs style, SV must include columns: strand_1, strand_2 (values '+' or '-'), and read_count.</p></dd>


<dt id="arg-svcol">SVcol<a class="anchor" aria-label="anchor" href="#arg-svcol"></a></dt>
<dd><p>Default is TRUE. Colour SVs or not</p></dd>


<dt id="arg-svalpha">svalpha<a class="anchor" aria-label="anchor" href="#arg-svalpha"></a></dt>
<dd><p>the alpha scaling of the SV lines, default = 0.5</p></dd>


<dt id="arg-svwidth">svwidth<a class="anchor" aria-label="anchor" href="#arg-svwidth"></a></dt>
<dd><p>Width of SV width curves, default = 1.0</p></dd>


<dt id="arg-adj">adj<a class="anchor" aria-label="anchor" href="#arg-adj"></a></dt>
<dd><p>adjustment for gene labels</p></dd>


<dt id="arg-genes">genes<a class="anchor" aria-label="anchor" href="#arg-genes"></a></dt>
<dd><p>vector of genes to annotate, will add a dashed vertical line and label</p></dd>


<dt id="arg-tickwidth">tickwidth<a class="anchor" aria-label="anchor" href="#arg-tickwidth"></a></dt>
<dd><p>Spacing of ticks (in Mb) when only 1 chromosome is plotted</p></dd>


<dt id="arg-chrstart">chrstart<a class="anchor" aria-label="anchor" href="#arg-chrstart"></a></dt>
<dd><p>Start of region (in Mb) when plotting a single chromosome</p></dd>


<dt id="arg-chrend">chrend<a class="anchor" aria-label="anchor" href="#arg-chrend"></a></dt>
<dd><p>End of region (in Mb) when plotting a single chromosome</p></dd>


<dt id="arg-shape">shape<a class="anchor" aria-label="anchor" href="#arg-shape"></a></dt>
<dd><p>shape for plotting, default = 16</p></dd>


<dt id="arg-positionticks">positionticks<a class="anchor" aria-label="anchor" href="#arg-positionticks"></a></dt>
<dd><p>set to TRUE to use position ticks rather than chromosome ticks</p></dd>


<dt id="arg-ideogram">ideogram<a class="anchor" aria-label="anchor" href="#arg-ideogram"></a></dt>
<dd><p>plot ideogram at the top, default = TRUE</p></dd>


<dt id="arg-sv-style">sv_style<a class="anchor" aria-label="anchor" href="#arg-sv-style"></a></dt>
<dd><p>visualization style for structural variants: "curves" (default, bezier curves), "lines_and_arcs" (colored lines with arcs), or "both"</p></dd>


<dt id="arg-sv-arc-alpha">sv_arc_alpha<a class="anchor" aria-label="anchor" href="#arg-sv-arc-alpha"></a></dt>
<dd><p>transparency of SV arcs when using lines_and_arcs style, default = 0.3</p></dd>


<dt id="arg-sv-arc-height">sv_arc_height<a class="anchor" aria-label="anchor" href="#arg-sv-arc-height"></a></dt>
<dd><p>height of SV arcs as fraction of maxCN when using lines_and_arcs style, default = 0.2</p></dd>


<dt id="arg-show-sv-read-axis">show_sv_read_axis<a class="anchor" aria-label="anchor" href="#arg-show-sv-read-axis"></a></dt>
<dd><p>show secondary y-axis for SV read support when using lines_and_arcs style, default = TRUE. Works with both identity and squashy y-axis transformations.</p></dd>


<dt id="arg-sv-read-axis-scale">sv_read_axis_scale<a class="anchor" aria-label="anchor" href="#arg-sv-read-axis-scale"></a></dt>
<dd><p>maximum value for SV read support axis (auto-scaled if NULL), default = NULL</p></dd>


<dt id="arg-show-sv-legend">show_sv_legend<a class="anchor" aria-label="anchor" href="#arg-show-sv-legend"></a></dt>
<dd><p>show legend for SV orientations when using lines_and_arcs style, default = FALSE</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>ggplot2 plot</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">plotCNprofile</span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making CN profile and BAF plot for cell - SA921-A90554A-R03-C44</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Making fixed-width bins for bin size 500000 ...</span>
<span class="r-plt img"><img src="plotCNprofile-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer></div>






  </body></html>

