<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Heatmap plot — plotHeatmap • schnapps</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap plot — plotHeatmap"><meta property="og:description" content="Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">schnapps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.6.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for schnapps</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heatmap plot</h1>
    
    <div class="hidden name"><code>plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">plotHeatmap</span><span class="op">(</span>
  <span class="va">cn</span>,
  tree <span class="op">=</span> <span class="cn">NULL</span>,
  clusters <span class="op">=</span> <span class="cn">NULL</span>,
  normalize_ploidy <span class="op">=</span> <span class="cn">FALSE</span>,
  normalize_tree <span class="op">=</span> <span class="cn">FALSE</span>,
  branch_length <span class="op">=</span> <span class="fl">1</span>,
  spacer_cols <span class="op">=</span> <span class="fl">20</span>,
  plottree <span class="op">=</span> <span class="cn">TRUE</span>,
  plotcol <span class="op">=</span> <span class="st">"state"</span>,
  reorderclusters <span class="op">=</span> <span class="cn">FALSE</span>,
  pctcells <span class="op">=</span> <span class="fl">0.05</span>,
  library_mapping <span class="op">=</span> <span class="cn">NULL</span>,
  clone_pal <span class="op">=</span> <span class="cn">NULL</span>,
  sample_label_idx <span class="op">=</span> <span class="fl">1</span>,
  fillna <span class="op">=</span> <span class="cn">TRUE</span>,
  frequencycutoff <span class="op">=</span> <span class="fl">2</span>,
  maxf <span class="op">=</span> <span class="cn">NULL</span>,
  plotfrequency <span class="op">=</span> <span class="cn">FALSE</span>,
  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,
  show_library_label <span class="op">=</span> <span class="cn">TRUE</span>,
  show_clone_label <span class="op">=</span> <span class="cn">TRUE</span>,
  show_clone_text <span class="op">=</span> <span class="cn">TRUE</span>,
  widenarm <span class="op">=</span> <span class="cn">FALSE</span>,
  umapmetric <span class="op">=</span> <span class="st">"euclidean"</span>,
  chrlabels <span class="op">=</span> <span class="cn">TRUE</span>,
  SV <span class="op">=</span> <span class="cn">NULL</span>,
  seed <span class="op">=</span> <span class="cn">NULL</span>,
  nticks <span class="op">=</span> <span class="fl">4</span>,
  fillgenome <span class="op">=</span> <span class="cn">FALSE</span>,
  annotation_height <span class="op">=</span> <span class="cn">NULL</span>,
  annofontsize <span class="op">=</span> <span class="fl">14</span>,
  na_col <span class="op">=</span> <span class="st">"white"</span>,
  linkheight <span class="op">=</span> <span class="fl">5</span>,
  newlegendname <span class="op">=</span> <span class="cn">NULL</span>,
  str_to_remove <span class="op">=</span> <span class="cn">NULL</span>,
  maxCNcol <span class="op">=</span> <span class="fl">11</span>,
  anno_width <span class="op">=</span> <span class="fl">0.4</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>cn</dt>
<dd><p>Either a hscn object or a single cell allele specific copy number dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `state`, `copy`</p></dd>
<dt>tree</dt>
<dd><p>Tree in newick format to plot alongside the heatmap, default = NULL</p></dd>
<dt>normalize_ploidy</dt>
<dd><p>Normalize ploidy of all cells to 2</p></dd>
<dt>normalize_tree</dt>
<dd><p>default = FALSE</p></dd>
<dt>branch_length</dt>
<dd><p>scales branch lengths to this size, default = 2</p></dd>
<dt>spacer_cols</dt>
<dd><p>number of empty columns between chromosomes, default = 20</p></dd>
<dt>plottree</dt>
<dd><p>Binary value of whether to plot tree or not, default = TRUE</p></dd>
<dt>plotcol</dt>
<dd><p>Which column to colour the heatmap by, should be one of "state", "state_BAF", "state_phase", "state_AS", "state_min", "copy", "BAF", "Min", "Maj"</p></dd>
<dt>pctcells</dt>
<dd><p>Minimum size of cluster in terms of perecentage of cells in umap clustering</p></dd>
<dt>library_mapping</dt>
<dd><p>Named vector mapping library names to labels for legend</p></dd>
<dt>clone_pal</dt>
<dd><p>pallette to colour clusters by</p></dd>
<dt>sample_label_idx</dt>
<dd><p>default = 1</p></dd>
<dt>fillna</dt>
<dd><p>Smooth over NA values, default = TRUE</p></dd>
<dt>frequencycutoff</dt>
<dd><p>default = 2</p></dd>
<dt>maxf</dt>
<dd><p>Max frequency when plotting the frequency track, default = NULL infers this from the data</p></dd>
<dt>plotfrequency</dt>
<dd><p>Plot the frequency track of gains and losses across the genome</p></dd>
<dt>show_legend</dt>
<dd><p>plot legend or not, boolean</p></dd>
<dt>show_library_label</dt>
<dd><p>show library label or not, boolean</p></dd>
<dt>show_clone_label</dt>
<dd><p>show clone label or not, boolean</p></dd>
<dt>umapmetric</dt>
<dd><p>metric to use in umap dimensionality reduction if no clusters are specified</p></dd>
<dt>chrlabels</dt>
<dd><p>include chromosome labels or not, boolean</p></dd>
<dt>SV</dt>
<dd><p>sv data frame</p></dd>
<dt>seed</dt>
<dd><p>seed for UMAP</p></dd>
<dt>nticks</dt>
<dd><p>number of ticks in x-axis label when plotting a single chromosome</p></dd>
<dt>fillgenome</dt>
<dd><p>fill in any missing bins and add NA to centromeric regions</p></dd>
<dt>na_col</dt>
<dd><p>colour of NA values</p></dd>
<dt>linkheight</dt>
<dd><p>height of x-axis ticks</p></dd>
<dt>newlegendname</dt>
<dd><p>overwrite default legend name</p></dd>
<dt>str_to_remove</dt>
<dd><p>string to remove from cell_id's when plotting labels</p></dd>
<dt>maxCNcol</dt>
<dd><p>max value for color scale when plotting raw data</p></dd>
<dt>anno_width</dt>
<dd><p>width of left annotations</p>
<p>If clusters are set to NULL then the function will compute clusters using UMAP and HDBSCAN.</p>
<p>#'</p></dd>
<dt>cluster</dt>
<dd><p>data.frame assigning cells to clusters, needs the following columns `cell_id`, `clone_id` default = NULL</p></dd>
<dt>reordercluster</dt>
<dd><p>Reorder the cells according to cluster if no tree is specified</p></dd>
</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

