<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Heatmap plot — plotHeatmap • schnapps</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Heatmap plot — plotHeatmap" />
<meta property="og:description" content="Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">schnapps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.6.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for schnapps</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Heatmap plot</h1>
    
    <div class="hidden name"><code>plotHeatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot a heatmap where rows are cells, columns are genome coordinates and colours map to (allele-specific) copy-number states</p>
    </div>

    <pre class="usage"><span class='fu'>plotHeatmap</span><span class='op'>(</span>
  <span class='va'>cn</span>,
  tree <span class='op'>=</span> <span class='cn'>NULL</span>,
  clusters <span class='op'>=</span> <span class='cn'>NULL</span>,
  normalize_ploidy <span class='op'>=</span> <span class='cn'>FALSE</span>,
  normalize_tree <span class='op'>=</span> <span class='cn'>FALSE</span>,
  branch_length <span class='op'>=</span> <span class='fl'>1</span>,
  spacer_cols <span class='op'>=</span> <span class='fl'>20</span>,
  plottree <span class='op'>=</span> <span class='cn'>TRUE</span>,
  plotcol <span class='op'>=</span> <span class='st'>"state"</span>,
  reorderclusters <span class='op'>=</span> <span class='cn'>FALSE</span>,
  pctcells <span class='op'>=</span> <span class='fl'>0.05</span>,
  library_mapping <span class='op'>=</span> <span class='cn'>NULL</span>,
  clone_pal <span class='op'>=</span> <span class='cn'>NULL</span>,
  sample_label_idx <span class='op'>=</span> <span class='fl'>1</span>,
  fillna <span class='op'>=</span> <span class='cn'>TRUE</span>,
  frequencycutoff <span class='op'>=</span> <span class='fl'>2</span>,
  maxf <span class='op'>=</span> <span class='cn'>NULL</span>,
  plotfrequency <span class='op'>=</span> <span class='cn'>FALSE</span>,
  show_legend <span class='op'>=</span> <span class='cn'>TRUE</span>,
  show_library_label <span class='op'>=</span> <span class='cn'>TRUE</span>,
  show_clone_label <span class='op'>=</span> <span class='cn'>TRUE</span>,
  widenarm <span class='op'>=</span> <span class='cn'>FALSE</span>,
  umapmetric <span class='op'>=</span> <span class='st'>"euclidean"</span>,
  chrlabels <span class='op'>=</span> <span class='cn'>TRUE</span>,
  SV <span class='op'>=</span> <span class='cn'>NULL</span>,
  seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  nticks <span class='op'>=</span> <span class='fl'>4</span>,
  fillgenome <span class='op'>=</span> <span class='cn'>FALSE</span>,
  annotation_height <span class='op'>=</span> <span class='cn'>NULL</span>,
  annofontsize <span class='op'>=</span> <span class='fl'>14</span>,
  na_col <span class='op'>=</span> <span class='st'>"white"</span>,
  linkheight <span class='op'>=</span> <span class='fl'>5</span>,
  newlegendname <span class='op'>=</span> <span class='cn'>NULL</span>,
  str_to_remove <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>...</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>cn</th>
      <td><p>Either a hscn object or a single cell allele specific copy number dataframe with the following columns: `cell_id`, `chr`, `start`, `end`, `state`, `copy`</p></td>
    </tr>
    <tr>
      <th>tree</th>
      <td><p>Tree in newick format to plot alongside the heatmap, default = NULL</p></td>
    </tr>
    <tr>
      <th>normalize_ploidy</th>
      <td><p>Normalize ploidy of all cells to 2</p></td>
    </tr>
    <tr>
      <th>normalize_tree</th>
      <td><p>default = FALSE</p></td>
    </tr>
    <tr>
      <th>branch_length</th>
      <td><p>scales branch lengths to this size, default = 2</p></td>
    </tr>
    <tr>
      <th>spacer_cols</th>
      <td><p>number of empty columns between chromosomes, default = 20</p></td>
    </tr>
    <tr>
      <th>plottree</th>
      <td><p>Binary value of whether to plot tree or not, default = TRUE</p></td>
    </tr>
    <tr>
      <th>plotcol</th>
      <td><p>Which column to colour the heatmap by, should be one of "state", "state_BAF", "state_phase", "state_AS", "state_min", "copy", "BAF", "Min", "Maj"</p></td>
    </tr>
    <tr>
      <th>pctcells</th>
      <td><p>Minimum size of cluster in terms of perecentage of cells in umap clustering</p></td>
    </tr>
    <tr>
      <th>library_mapping</th>
      <td><p>Named vector mapping library names to labels for legend</p></td>
    </tr>
    <tr>
      <th>clone_pal</th>
      <td><p>pallette to colour clusters by</p></td>
    </tr>
    <tr>
      <th>sample_label_idx</th>
      <td><p>default = 1</p></td>
    </tr>
    <tr>
      <th>fillna</th>
      <td><p>Smooth over NA values, default = TRUE</p></td>
    </tr>
    <tr>
      <th>frequencycutoff</th>
      <td><p>default = 2</p></td>
    </tr>
    <tr>
      <th>maxf</th>
      <td><p>Max frequency when plotting the frequency track, default = NULL infers this from the data</p></td>
    </tr>
    <tr>
      <th>plotfrequency</th>
      <td><p>Plot the frequency track of gains and losses across the genome</p></td>
    </tr>
    <tr>
      <th>show_legend</th>
      <td><p>plot legend or not, boolean</p></td>
    </tr>
    <tr>
      <th>show_library_label</th>
      <td><p>show library label or not, boolean</p></td>
    </tr>
    <tr>
      <th>show_clone_label</th>
      <td><p>show clone label or not, boolean</p></td>
    </tr>
    <tr>
      <th>umapmetric</th>
      <td><p>metric to use in umap dimensionality reduction if no clusters are specified</p></td>
    </tr>
    <tr>
      <th>chrlabels</th>
      <td><p>include chromosome labels or not, boolean</p></td>
    </tr>
    <tr>
      <th>SV</th>
      <td><p>sv data frame</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>seed for UMAP</p></td>
    </tr>
    <tr>
      <th>nticks</th>
      <td><p>number of ticks in x-axis label when plotting a single chromosome</p></td>
    </tr>
    <tr>
      <th>fillgenome</th>
      <td><p>fill in any missing bins and add NA to centromeric regions</p></td>
    </tr>
    <tr>
      <th>na_col</th>
      <td><p>colour of NA values</p></td>
    </tr>
    <tr>
      <th>linkheight</th>
      <td><p>height of x-axis ticks</p></td>
    </tr>
    <tr>
      <th>newlegendname</th>
      <td><p>overwrite default legend name</p></td>
    </tr>
    <tr>
      <th>str_to_remove</th>
      <td><p>string to remove from cell_id's when plotting labels</p>
<p>If clusters are set to NULL then the function will compute clusters using UMAP and HDBSCAN.</p>
<p>#'</p></td>
    </tr>
    <tr>
      <th>cluster</th>
      <td><p>data.frame assigning cells to clusters, needs the following columns `cell_id`, `clone_id` default = NULL</p></td>
    </tr>
    <tr>
      <th>reordercluster</th>
      <td><p>Reorder the cells according to cluster if no tree is specified</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


