<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Allele Specific Copy Number Inference in scRNAseq • signals</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Allele Specific Copy Number Inference in scRNAseq">
<meta property="og:description" content="signals">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Allele Specific Copy Number Inference in
scRNAseq</h1>
            
      
      
      <div class="hidden name"><code>ASCN-RNA.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">signals</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This vignette illustrates how to perform allele specific copy number
inference in scRNAseq data. scRNAseq data is much more sparse than
scDNAseq, so we perform inference on a chromosome arm basis by merging
counts across segments or chromosome arms.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>The data needed to perform the allele specific copy number inference
are SNP calls of heterozygous positions in individual single cells.
Ideally haplotype block calling should also be performed, so that each
allele in each cell can be assigned to a haplotype block. To get this
data you will ideally need a whole genome sequenced bulk sample of
normal tissue to identify heterozygous SNPs and then you can use a tool
such as cellSNP or vartrix to get per cell counts. An example dataset
from ~814 cells is included here. For this dataset we also have paired
scDNAseq, where we have called HSCN using <code>signals</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"haplotypes_rna"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scdna">scDNA<a class="anchor" aria-label="anchor" href="#scdna"></a>
</h2>
<p>We can use information from scDNAseq to phase the SNPs in scRNAseq.
To do this we’ll first compute haplotype specific copy number in
scDNAseq and use this phasing information in the scRNAseq.</p>
<p>It’s also possible to perform this phasing just based on scRNAseq,
but in general it will be less accurate.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"haplotypes"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CNbins"</span><span class="op">)</span></span>
<span><span class="va">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span></span>
<span><span class="va">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span><span class="op">)</span></span>
<span><span class="co">#&gt; VGLM    linear loop  1 :  loglikelihood = -313086.2337</span></span>
<span><span class="co">#&gt; VGLM    linear loop  2 :  loglikelihood = -309095.6185</span></span>
<span><span class="co">#&gt; VGLM    linear loop  3 :  loglikelihood = -308536.8714</span></span>
<span><span class="co">#&gt; VGLM    linear loop  4 :  loglikelihood = -308510.4127</span></span>
<span><span class="co">#&gt; VGLM    linear loop  5 :  loglikelihood = -308510.0155</span></span>
<span><span class="co">#&gt; VGLM    linear loop  6 :  loglikelihood = -308510.0119</span></span>
<span><span class="co">#&gt; VGLM    linear loop  7 :  loglikelihood = -308510.0118</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ascn-inference">ASCN inference<a class="anchor" aria-label="anchor" href="#ascn-inference"></a>
</h2>
<p>The first step is to format this dataframe. This will phase the
haplotypes across all the cells and add additional columns such as the
B-Allele Frequency. Our scDNAseq includes information on the phasing of
haplotype blocks which we’ll use as input.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"haplotypes_rna"</span><span class="op">)</span></span>
<span><span class="va">haplotypes_rna</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_rna</span><span class="op">(</span><span class="va">haplotypes_rna</span>, </span>
<span>                                    phased_haplotypes <span class="op">=</span> <span class="va">hscn</span><span class="op">$</span><span class="va">haplotype_phasing</span><span class="op">)</span></span></code></pre></div>
<p>As a first look at the data we can calculate the BAF per arm in each
cell and plot these distributions using <code>per_chr_baf_plot</code>.
Notice on this dataset that the BAFs in chromosomes <code>3p</code>,
<code>2p</code>, and <code>17</code> in almost all cells are very left
skewed towards 0 suggesting clonal LOH events in these chromosomes. We
can also see other chromosome arms such as 4p and 2q that have
multimodal distributions, suggesting different clusters of cells.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">per_chr_baf_plot</span><span class="op">(</span><span class="va">haplotypes_rna</span><span class="op">$</span><span class="va">block_counts</span>, perarm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<p>For comparison we can plot a similar distribution from our scDNAseq
dataset. Here we notice similar trends but note that data from the
scRNAseq is much more dispersed due to the lower number of counts per
cell. scRNAseq will have 1-2 orders of magnitude lower reads so our
ability to identify SNPs is reduced.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">per_chr_baf_plot</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span>, perarm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>We can also plot heatmaps of the 2 modalities as shown below, rows
are clustered by hierarchical clustering.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotHeatmapBAF</span><span class="op">(</span><span class="va">haplotypes_rna</span><span class="op">$</span><span class="va">block_counts</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotHeatmapBAF</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>We’ll now move on to inferring the allele specific copy number in the
scRNAseq. Rather than use chromosome arms, we first use the single cell
DNA sequencing data to identify segments that have different allele
specific copy number. We compute consensus copy number across all cells,
identify segments where the allele specific copy number phased state is
consistent and then finally merge any segments that are smaller than
10Mb.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">consensus</span> <span class="op">&lt;-</span> <span class="fu">consensuscopynumber</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">segments</span> <span class="op">&lt;-</span> <span class="fu">create_segments</span><span class="op">(</span><span class="va">consensus</span>, field <span class="op">=</span> <span class="st">"state_phase"</span><span class="op">)</span></span>
<span><span class="va">segments</span> <span class="op">&lt;-</span> <span class="fu">filter_segments</span><span class="op">(</span><span class="va">segments</span>, binwidth <span class="op">=</span> <span class="fl">10e6</span><span class="op">)</span></span></code></pre></div>
<p>Note, that if you have some other ways to obtain segments such as
bulk whole genome sequencing then that would also work. The
<code>segments</code> data frame needs to have the columns
<code>chr</code>, <code>start</code> and <code>end</code>.
Alternatively, if you don’t have any data that can be used to identify
segments a priori you can just use chromosome arms as the segmentation.
To do this just leave the segments option in any of the following
function empty and the functions will revert to the default of using
chromosome arms.</p>
<p>Our inference scheme is based on using dirichilet process clustering
using <a href="https://github.com/caravagn/VIBER" class="external-link">Viber</a> to cluster
cells that share similar BAF distributions across segments. We first
compute BAF frequencies across each segment in each cell.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haplotypes_per_segment</span> <span class="op">&lt;-</span> <span class="fu">per_segment_baf_mat</span><span class="op">(</span><span class="va">haplotypes_rna</span><span class="op">$</span><span class="va">block_counts</span>, segments <span class="op">=</span> <span class="va">segments</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ascn_rna_dp</span> <span class="op">&lt;-</span> <span class="fu">assign_states_dp</span><span class="op">(</span><span class="va">haplotypes_per_segment</span><span class="op">$</span><span class="va">bafpersegment</span>, <span class="va">haplotypes_rna</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 814   5</span></span>
<span><span class="co">#&gt; <span style="color: #BBBBBB; background-color: #0000BB;"> [ VIBER - variational fit ] </span></span></span>
<span><span class="co"><span style="color: #BBBBBB; background-color: #0000BB;">#&gt; </span></span></span>
<span><span class="va">ascn_rna_dp</span></span>
<span><span class="co">#&gt; Haplotype specific copy number object (RNA) </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; Number of cells: 814</span></span>
<span><span class="co">#&gt; Number of clusters: 2</span></span>
<span><span class="co">#&gt; Chromosomes used for clustering:</span></span></code></pre></div>
<p>Here, we found 2 clusters. We can visualize the separation of
clusters using a similar plot to before. Setting
<code>lableclones = TRUE</code> will colour each cell by their clone
assignment.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">per_segment_baf_plot</span><span class="op">(</span><span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">hscn</span>, labelclones <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-11-1.png" width="576"></p>
<div class="section level3">
<h3 id="heatmap">Heatmap<a class="anchor" aria-label="anchor" href="#heatmap"></a>
</h3>
<p>Finally we can plot heatmaps of the resulting outputs. Firstly using
the scRNA.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bins</span> <span class="op">&lt;-</span> <span class="fu">segments_to_bins</span><span class="op">(</span><span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">hscn</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">bins</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">clusters</span>,</span>
<span>            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            spacer_cols <span class="op">=</span> <span class="fl">15</span>,</span>
<span>            plotcol <span class="op">=</span> <span class="st">"BAF"</span>,</span>
<span>            plottree <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            show_library_label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            show_clone_label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            show_legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-13-1.png" width="816"></p>
<p>We can also do this for the scDNAseq data. For more straightforward
comparison we’ll perform the dirichlet process arm level inference as we
do for the scRNAseq. We can see that the scRNAseq has much lower
resolution that scDNAseq, but we are still able to recover the events
shared across all cells and clones that are at a reasonably high
frequency.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hscn_dna_segment</span> <span class="op">&lt;-</span> <span class="fu">per_segment_baf_mat</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span>, <span class="va">segments</span><span class="op">)</span></span>
<span><span class="va">ascn_dna_dp</span> <span class="op">&lt;-</span> <span class="fu">assign_states_dp</span><span class="op">(</span><span class="va">hscn_dna_segment</span><span class="op">$</span><span class="va">bafpersegment</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 250   5</span></span>
<span><span class="co">#&gt; <span style="color: #BBBBBB; background-color: #0000BB;"> [ VIBER - variational fit ] </span></span></span>
<span><span class="co"><span style="color: #BBBBBB; background-color: #0000BB;">#&gt; </span></span></span>
<span></span>
<span><span class="va">bins_dna</span> <span class="op">&lt;-</span> <span class="fu">segments_to_bins</span><span class="op">(</span><span class="va">ascn_dna_dp</span><span class="op">$</span><span class="va">hscn</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">bins_dna</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">ascn_dna_dp</span><span class="op">$</span><span class="va">clusters</span>,</span>
<span>            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            spacer_cols <span class="op">=</span> <span class="fl">15</span>,</span>
<span>            plotcol <span class="op">=</span> <span class="st">"BAF"</span>,</span>
<span>            plottree <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            show_library_label <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>            show_legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-14-1.png" width="816"></p>
</div>
</div>
<div class="section level2">
<h2 id="integration-with-seurat">Integration with Seurat<a class="anchor" aria-label="anchor" href="#integration-with-seurat"></a>
</h2>
<p>We can also integrate this inference with Seurat. We’ll first load a
gene expression counts matrix and then run the standard Seurat
pipeline,</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"countsmatrix"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">countsmatrix</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 814</span></span>
<span><span class="co">#&gt; Number of edges: 24960</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.7907</span></span>
<span><span class="co">#&gt; Number of communities: 7</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span>,  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></span></code></pre></div>
<p>Then we can use <code>createBAFassay</code> to add a <code>BAF</code>
assay and add <code>clone_id</code> to the metadata.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">createBAFassay</span><span class="op">(</span><span class="va">x</span>, <span class="va">ascn_rna_dp</span>, <span class="st">"hg19"</span><span class="op">)</span></span></code></pre></div>
<p>This means all the Seurat functionality is available to analyse the
allele specific copy number. For example we can plot low dimensional
embeddings of BAF values per cell per chromosome or coloured by clone as
shown below.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">x</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">segBAF</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "BAF-1-1-47000000"           "BAF-1-225500001-243700000" </span></span>
<span><span class="co">#&gt;  [3] "BAF-1-47000001-225500000"   "BAF-10-1-97000000"         </span></span>
<span><span class="co">#&gt;  [5] "BAF-10-97000001-130600000"  "BAF-11-119500001-130800000"</span></span>
<span><span class="co">#&gt;  [7] "BAF-11-1-20500000"          "BAF-11-20500001-82500000"  </span></span>
<span><span class="co">#&gt;  [9] "BAF-11-82500001-119500000"  "BAF-12-1-123000000"        </span></span>
<span><span class="co">#&gt; [11] "BAF-13-1-110300000"         "BAF-14-1-54000000"         </span></span>
<span><span class="co">#&gt; [13] "BAF-14-54000001-104000000"  "BAF-15-1-43000000"         </span></span>
<span><span class="co">#&gt; [15] "BAF-15-43000001-77000000"   "BAF-15-77000001-98500000"  </span></span>
<span><span class="co">#&gt; [17] "BAF-16-1-88700000"          "BAF-17-1-75300000"         </span></span>
<span><span class="co">#&gt; [19] "BAF-18-1-73100000"          "BAF-19-1-28000000"         </span></span>
<span><span class="co">#&gt; [21] "BAF-19-28000001-56300000"   "BAF-2-1-88500000"          </span></span>
<span><span class="co">#&gt; [23] "BAF-2-225000001-237300000"  "BAF-2-88500001-99000000"   </span></span>
<span><span class="co">#&gt; [25] "BAF-2-99000001-225000000"   "BAF-20-13000001-58400000"  </span></span>
<span><span class="co">#&gt; [27] "BAF-20-1-13000000"          "BAF-21-1-42600000"         </span></span>
<span><span class="co">#&gt; [29] "BAF-22-1-49400000"          "BAF-3-1-75500000"          </span></span>
<span><span class="co">#&gt; [31] "BAF-3-75500001-192300000"   "BAF-4-108500001-138500000" </span></span>
<span><span class="co">#&gt; [33] "BAF-4-138500001-176000000"  "BAF-4-176000001-187100000" </span></span>
<span><span class="co">#&gt; [35] "BAF-4-1-47500000"           "BAF-4-47500001-108500000"  </span></span>
<span><span class="co">#&gt; [37] "BAF-5-1-51000000"           "BAF-5-51000001-176600000"  </span></span>
<span><span class="co">#&gt; [39] "BAF-6-154000001-164500000"  "BAF-6-1-24000000"          </span></span>
<span><span class="co">#&gt; [41] "BAF-6-24000001-65500000"    "BAF-6-65500001-154000000"  </span></span>
<span><span class="co">#&gt; [43] "BAF-7-1-26000000"           "BAF-7-26000001-62000000"   </span></span>
<span><span class="co">#&gt; [45] "BAF-7-62000001-81500000"    "BAF-7-81500001-155100000"  </span></span>
<span><span class="co">#&gt; [47] "BAF-8-1-75000000"           "BAF-8-75000001-139900000"  </span></span>
<span><span class="co">#&gt; [49] "BAF-9-1-137400000"          "BAF-X-1-147100000"</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-3-75500001-192300000"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-10-1-97000000"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-2-99000001-225000000"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-2-99000001-225000000"</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"BAF"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>We can also pull out the allelic imbalance in individual genes. Let’s
look at some genes on 4p, where one group of cells is homozygous. Most
genes have very few counts, so we’ll pull out genes with a large number
of genotyped SNPs to take a look at.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chr4pgenes</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[[</span><span class="st">"gBAF"</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">meta.features</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">arm</span> <span class="op">==</span> <span class="st">"4p"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">totalsnpcounts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mygene</span> <span class="op">&lt;-</span> <span class="va">chr4pgenes</span><span class="op">$</span><span class="va">ensembl_gene_symbol</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"BAF-"</span>, <span class="va">mygene</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">mygene</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">x</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"BAF-"</span>, <span class="va">mygene</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"BAF"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>Note that even in this example where we have been able to identify
lots of SNP counts in this gene, most individual cells will still have
very few counts. This explains why in cluster C2, the violin plot shows
that there are lots of cells with BAF of 0.0 or 1.0, suggesting these
cells are homozygous. This however is likely just because the number of
counts is very small, even for a heterozygous SNP we’re very likely to
see cells where we have only sampled the reference or the alternate
allele. The key here, is that in cluster C1 we <em>only</em> see cells
with BAF = 0.0 which is unlikely to be down to sampling alone. In
contrast in C2 we see cells across a range of BAF, consistent with
expression of of both alleles + random sampling.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
