<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Allele Specific Copy Number Inference in scRNAseq • schnapps</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Allele Specific Copy Number Inference in scRNAseq">
<meta property="og:description" content="schnapps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">schnapps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for schnapps</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ASCN-RNA_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Allele Specific Copy Number Inference in scRNAseq</h1>
            
      
      
      <div class="hidden name"><code>ASCN-RNA.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">schnapps</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/">cowplot</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></code></pre></div>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>This vignette illustrates how to perform allele specific copy number inference in scRNAseq data. scRNAseq data is much more sparse than scDNAseq, so we perform inference on a chromosome arm basis by merging counts across arms.</p>
</div>
<div id="data" class="section level2">
<h2 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h2>
<p>The data needed to perform the allele specific copy number inference are SNP calls of heterozygous positions in individual single cells. Ideally haplotype block calling should also be performed, so that each allele in each cell can be assigned to a haplotype block. To get this data you will ideally need a whole genome sequenced bulk sample of normal tissue to identify heterozygous SNPs and then you can use a tool such as cellSNP or vartrix to get per cell counts. An example dataset from ~814 cells is included here. For this dataset we also have paired scDNAseq, where we have called HSCN using <code>schnapps</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"haplotypes_rna"</span><span class="op">)</span></code></pre></div>
</div>
<div id="scdna" class="section level2">
<h2 class="hasAnchor">
<a href="#scdna" class="anchor"></a>scDNA</h2>
<p>We can use information from scDNAseq to phase the SNPs in scRNAseq. To do this we’ll first compute haplotype specific copy number in scDNAseq and use this phasing information in the scRNAseq.</p>
<p>It’s also possible to perform this phasing just based on scRNAseq, but in general it will be less accurate.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"haplotypes"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"CNbins"</span><span class="op">)</span>
<span class="va">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span>
<span class="va">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span><span class="op">)</span></code></pre></div>
</div>
<div id="ascn-inference" class="section level2">
<h2 class="hasAnchor">
<a href="#ascn-inference" class="anchor"></a>ASCN inference</h2>
<p>The first step is to format this dataframe. This will phase the haplotypes across all the cells and add additional columns such as the B-Allele Frequency. Our scDNAseq includes information on the phasing of haplotype blocks which we’ll use as input.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"haplotypes_rna"</span><span class="op">)</span>
<span class="va">haplotypes_rna</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_rna</span><span class="op">(</span><span class="va">haplotypes_rna</span>, 
                                    phased_haplotypes <span class="op">=</span> <span class="va">hscn</span><span class="op">$</span><span class="va">haplotype_phasing</span><span class="op">)</span></code></pre></div>
<p>As a first look at the data we can calculate the BAF per arm in each cell and plot these distributions using <code>per_chr_baf</code>. Notice on this dataset that the BAFs in chromosomes <code>3p</code>, <code>2p</code>, and <code>17</code> in almost all cells are very left skewed towards 0 suggesting clonal LOH events in these chromosomes. We can also see other chromosome arms such as 4p and 2q that have multimodal distributions, suggesting different clusters of cells.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">per_chr_baf</span><span class="op">(</span><span class="va">haplotypes_rna</span>, perarm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-5-1.png" width="576"></p>
<p>For comparison we can plot a similar distribution from our scDNAseq dataset. Here we notice similar trends but note that data from the scRNAseq is much more dispersed due to the lower number of counts per cell. scRNAseq will have 1-2 orders of magnitude lower reads so our ability to identify SNPs is reduced.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">per_chr_baf</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span>, perarm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
<p>We can also plot heatmaps of the 2 modalities as shown below.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotHeatmapBAF</span><span class="op">(</span><span class="va">haplotypes_rna</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-7-1.png" width="672"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">plotHeatmapBAF</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-8-1.png" width="672"></p>
<p>We’ll now move on to inferring the allele specific copy number in the scRNAseq. We use dirichilet process clustering using <a href="https://github.com/caravagn/VIBER">Viber</a> to cluster cells that share similar BAF distributions across chromosomes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haplotypes_per_arm</span> <span class="op">&lt;-</span> <span class="fu">per_arm_baf_mat</span><span class="op">(</span><span class="va">haplotypes_rna</span><span class="op">)</span>

<span class="va">ascn_rna_dp</span> <span class="op">&lt;-</span> <span class="fu">assign_states_dp</span><span class="op">(</span><span class="va">haplotypes_per_arm</span><span class="op">$</span><span class="va">bafperchr</span><span class="op">)</span>
<span class="co">#&gt; [1] 814   5</span>
<span class="co">#&gt;  [ VIBER - variational fit ]</span>
<span class="va">ascn_rna_dp</span>
<span class="co">#&gt; Haplotype specific copy number object (RNA) </span>
<span class="co">#&gt;  </span>
<span class="co">#&gt; Number of cells: 814</span>
<span class="co">#&gt; Number of clusters: 2</span>
<span class="co">#&gt; Chromosomes used for clustering: chr20p, chr16p, chr9p, chr4p, chr18</span></code></pre></div>
<p>Here, we found 2 clusters. We can visualize the separation of clusters using the same plot as before.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">per_chr_baf</span><span class="op">(</span><span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">hscn</span>, perarm <span class="op">=</span> <span class="cn">TRUE</span>, labelclones <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-10-1.png" width="576"></p>
<div id="heatmap" class="section level3">
<h3 class="hasAnchor">
<a href="#heatmap" class="anchor"></a>Heatmap</h3>
<p>Finally we can plot heatmaps of the resulting outputs. Firstly using the scRNA.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">hscn</span>, 
            clusters <span class="op">=</span> <span class="va">ascn_rna_dp</span><span class="op">$</span><span class="va">clusters</span>,
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>,
            spacer_cols <span class="op">=</span> <span class="fl">15</span>, 
            plotcol <span class="op">=</span> <span class="st">"BAF"</span>, 
            plottree <span class="op">=</span> <span class="cn">FALSE</span>,
            widenarm <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#make sure this is true</span>
            show_legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-11-1.png" width="816"></p>
<p>We can also do this for the scDNAseq data. For more straightforward comparison we’ll perform the dirichilet process arm level inference as we do for the scRNAseq. We can see that the scRNAseq has much lower resolution that scDNAseq, but we are still able to recover the events shared across all cells and clones that are at a reasonably high frequency.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hscn_dna_arm</span> <span class="op">&lt;-</span> <span class="fu">per_arm_baf_mat</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>
<span class="va">ascn_dna_dp</span> <span class="op">&lt;-</span> <span class="fu">assign_states_dp</span><span class="op">(</span><span class="va">hscn_dna_arm</span><span class="op">$</span><span class="va">bafperchr</span><span class="op">)</span>
<span class="co">#&gt; [1] 250   5</span>
<span class="co">#&gt;  [ VIBER - variational fit ]</span>
<span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">ascn_dna_dp</span><span class="op">$</span><span class="va">hscn</span>,
            clusters <span class="op">=</span> <span class="va">ascn_dna_dp</span><span class="op">$</span><span class="va">clusters</span>,
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>,
            spacer_cols <span class="op">=</span> <span class="fl">15</span>, 
            plotcol <span class="op">=</span> <span class="st">"BAF"</span>, 
            plottree <span class="op">=</span> <span class="cn">FALSE</span>,
            widenarm <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#make sure this is true</span>
            show_legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-12-1.png" width="816"></p>
</div>
</div>
<div id="integration-with-seurat" class="section level2">
<h2 class="hasAnchor">
<a href="#integration-with-seurat" class="anchor"></a>Integration with Seurat</h2>
<p>We can also integrate this inference with Seurat. We’ll first load a gene expression counts matrix and then run the standard Seurat pipeline,</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"countsmatrix"</span><span class="op">)</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat">Seurat</a></span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html">CreateSeuratObject</a></span><span class="op">(</span><span class="va">countsmatrix</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html">NormalizeData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html">FindVariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html">RunPCA</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html">FindClusters</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span><span class="op">)</span>
<span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of nodes: 814</span>
<span class="co">#&gt; Number of edges: 24960</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Running Louvain algorithm...</span>
<span class="co">#&gt; Maximum modularity in 10 random starts: 0.7907</span>
<span class="co">#&gt; Number of communities: 7</span>
<span class="co">#&gt; Elapsed time: 0 seconds</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">x</span>,  dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"pca"</span><span class="op">)</span></code></pre></div>
<p>Then we can use <code>createBAFassay</code> to add a <code>BAF</code> assay and add <code>clone_id</code> to the metadata.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">createBAFassay</span><span class="op">(</span><span class="va">x</span>, <span class="va">ascn_rna_dp</span><span class="op">)</span></code></pre></div>
<p>This means all the Seurat functionality is available to analyse the allele specific copy number. For example we can plot low dimensional embeddings of BAF values per cell per chromosome or coloured by clone as shown below.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">x</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-3q"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"BAF-10p"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_color_gradientn</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu">scBAFstate_cols</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-17-1.png" width="700"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">x</span>,<span class="st">"BAF-10p"</span>, group.by <span class="op">=</span> <span class="st">"DP_cloneid"</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN-RNA_files/figure-html/unnamed-chunk-18-1.png" width="700"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
