<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Allele Specific Copy Number Inference • signals</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Allele Specific Copy Number Inference">
<meta property="og:description" content="signals">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ASCN_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Allele Specific Copy Number Inference</h1>
            
      
      
      <div class="hidden name"><code>ASCN.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">signals</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This vignette illustrates how to perform allele specific copy number inference with DLP+ data. However the method should be generally applicable to any single cell copy number method. The input requirements are discussed below.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>The data needed to perform the allele specific copy number inference are binned copy number data and cell specific haplotype calls. The copy number data and the haplotype data should use the same bins.</p>
<p>A small test data set of 250 cells is provided with the package, we’ll load this data and inspect these 2 dataframe.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">haplotypes</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span>
<span class="co">#&gt;   chr   start     end reads     copy state               cell_id</span>
<span class="co">#&gt; 1   1 2000001 2500000   536 2.960868     2 SA921-A90554A-R03-C44</span>
<span class="co">#&gt; 2   1 3000001 3500000   499 2.603584     2 SA921-A90554A-R03-C44</span>
<span class="co">#&gt; 3   1 4000001 4500000   569 2.216388     2 SA921-A90554A-R03-C44</span>
<span class="co">#&gt; 4   1 4500001 5000000   524 2.039672     2 SA921-A90554A-R03-C44</span>
<span class="co">#&gt; 5   1 5000001 5500000   576 2.210792     2 SA921-A90554A-R03-C44</span>
<span class="co">#&gt; 6   1 5500001 6000000   516 2.028686     2 SA921-A90554A-R03-C44</span></code></pre></div>
<p>The haplotype data comes in long format, but the package requires data in wide format so we will use the helper function <code>format_haplotypes_dlp</code> to convert the data to the correct format, this function will also remove any bins present in the haplotypes data frame that is not present in the CNbins data frame.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">haplotypes</span><span class="op">)</span>
<span class="co">#&gt;                  cell_id chr   start     end hap_label allele1 allele0</span>
<span class="co">#&gt; 1 SA1090-A96213A-R20-C13   1 2000001 2500000       101       0       3</span>
<span class="co">#&gt; 2 SA1090-A96213A-R20-C13   1 2000001 2500000       104       0       1</span>
<span class="co">#&gt; 3 SA1090-A96213A-R20-C13   1 2000001 2500000       106       0       3</span>
<span class="co">#&gt; 4 SA1090-A96213A-R20-C13   1 2000001 2500000       108       0       4</span>
<span class="co">#&gt; 5 SA1090-A96213A-R20-C13   1 2000001 2500000       112       0      15</span>
<span class="co">#&gt; 6 SA1090-A96213A-R20-C13   1 2000001 2500000       113       0       1</span>
<span class="co">#&gt;   totalcounts</span>
<span class="co">#&gt; 1           3</span>
<span class="co">#&gt; 2           1</span>
<span class="co">#&gt; 3           3</span>
<span class="co">#&gt; 4           4</span>
<span class="co">#&gt; 5          15</span>
<span class="co">#&gt; 6           1</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="inference">Inference<a class="anchor" aria-label="anchor" href="#inference"></a>
</h2>
<p>The output of <code>signals</code> is allele specific states per bin. <code>signals</code> has 2 options to perform this inference, the first, <code>callAlleleSpecificCN</code>, infers allele specific states using mirrored B allele frequencies as is typically done in bulk whole genome sequencing. This infers states of the form <code>A|B</code> where <code>B &lt; A</code>. The second option, <code>callHaplotypeSpecificCN</code> takes into account the possibility that different alleles may be lost or gained in different cells. In order to identify these events, <code>signals</code> clusters cells and infers the haplotype phase based on regions within clusters that have large allelic imbalance. The output for this method therefore allows for <code>B&gt;A</code>.</p>
<div class="section level3">
<h3 id="callallelespecificcn">
<code>callAlleleSpecificCN</code><a class="anchor" aria-label="anchor" href="#callallelespecificcn"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ascn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callAlleleSpecificCN.html">callAlleleSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span><span class="op">)</span></code></pre></div>
<p>As a sanity check we can plot the B allele frequency (BAF) as a function of inferred state. We would expect to see a distribution around the expected value which is exactly what we see below.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBAFperstate.html">plotBAFperstate</a></span><span class="op">(</span><span class="va">ascn</span>, maxstate <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<p>Now we can plot the copy number heatmaps. First we’ll plot the heatmap based on copy number states. We can provide the <code>plotHeatmap</code> function with a tree in newick format to plot an ordered heatmap and/or a dataframe with cluster assignments. By default the <code>plotHeatmap</code> function does some clustering using hdbscan if neither of these is provided.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cl</span> <span class="op">&lt;-</span> <span class="fu">umap_clustering</span><span class="op">(</span><span class="va">ascn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">ascn</span>, 
            plotcol <span class="op">=</span> <span class="st">"state"</span>, 
            clusters <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span>,
            tree <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">tree</span>, 
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#order according to tree</span>
            plottree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#do not plot the tree but keep the ordering )</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<p>Now we can plot some heatmaps to visualize the allele specific state. This heatmap plots the minor copy number state.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">ascn</span>, plotcol <span class="op">=</span> <span class="st">"Min"</span>,
            clusters <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span>,
            tree <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">tree</span>, 
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#order according to tree</span>
            plottree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#do not plot the tree but keep the ordering )</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-9-1.png" width="960"></p>
<p>Discrete representation of allelic imbalance:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">ascn</span>, 
            plotcol <span class="op">=</span> <span class="st">"state_phase"</span>, 
            clusters <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span>,
            tree <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">tree</span>, 
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#order according to tree</span>
            plottree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#do not plot the tree but keep the ordering )</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-10-1.png" width="960"></p>
<p>There are also some function to plot per cell BAF and state plots. You can specify a cell, or the default will take the first cell in the data frame.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCNprofileBAF.html">plotCNprofileBAF</a></span><span class="op">(</span><span class="va">ascn</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-11-1.png" width="960"></p>
<p>Here we’ll plot an equivalent plot merging data across cell clusters using the utility function <code>consensyscopynumber</code>. Here, the <code>cell_id</code> column becomes the</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">consensus_clusters</span> <span class="op">&lt;-</span> <span class="fu">consensuscopynumber</span><span class="op">(</span><span class="va">ascn</span><span class="op">$</span><span class="va">data</span>, cl <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotCNprofileBAF.html">plotCNprofileBAF</a></span><span class="op">(</span><span class="va">consensus_clusters</span>, cellid <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-12-1.png" width="960"></p>
<p>The object returned by the inference function is of type <code>ascn</code>, if you want to pull out the data to use in your own scripts you can just do <code>ascn$data</code>. This will return a <code>data.frame</code> similar to the input data frames but with additional columns specifying the allele specific states.</p>
</div>
<div class="section level3">
<h3 id="callhaplotypespecificcn">
<code>callHaplotypeSpecificCN</code><a class="anchor" aria-label="anchor" href="#callhaplotypespecificcn"></a>
</h3>
<p>Now we’ll do the analysis that phases haplotypes across cells.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span><span class="op">)</span></code></pre></div>
<p>We’ll now generate the same plots as before. We can see in the BAF per state plot that we now observe states with BAF &gt; 0.5 meaning that we see different alleles gained or lost in different cells.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotBAFperstate.html">plotBAFperstate</a></span><span class="op">(</span><span class="va">hscn</span>, maxstate <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-14-1.png" width="960"></p>
<p>In the allelic imbalance heatmap we can see that some cells have gained the “B” allele and others the “A” allele resulting in imbalance spanning from 0 to 1.</p>
<p>In the discrete version of this plot we see similar patterns.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">hscn</span>, plotcol <span class="op">=</span> <span class="st">"state_phase"</span>,
            clusters <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span>,
            tree <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">tree</span>, 
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#order according to tree</span>
            plottree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#do not plot the tree but keep the ordering </span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-15-1.png" width="960"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">hscn</span>, plotcol <span class="op">=</span> <span class="st">"Min"</span>,
            clusters <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span>,
            tree <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">tree</span>, 
            reorderclusters <span class="op">=</span> <span class="cn">TRUE</span>, <span class="co">#order according to tree</span>
            plottree <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="co">#do not plot the tree but keep the ordering </span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-16-1.png" width="960"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCNprofileBAF.html">plotCNprofileBAF</a></span><span class="op">(</span><span class="va">hscn</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-17-1.png" width="960"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">consensus_clusters</span> <span class="op">&lt;-</span> <span class="fu">consensuscopynumber</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span>, cl <span class="op">=</span> <span class="va">cl</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span>

<span class="fu"><a href="../reference/plotCNprofileBAF.html">plotCNprofileBAF</a></span><span class="op">(</span><span class="va">consensus_clusters</span>, cellid <span class="op">=</span> <span class="st">"B"</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
<p>An alternative plotting style is to plot the copy number of the individual copy number.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCNprofileBAF.html">plotCNprofileBAF</a></span><span class="op">(</span><span class="va">consensus_clusters</span>, cellid <span class="op">=</span> <span class="st">"C"</span>, homolog <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p><img src="ASCN_files/figure-html/unnamed-chunk-19-1.png" width="960"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
