<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Basic scDNAseq analysis • signals</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Basic scDNAseq analysis">
<meta property="og:description" content="signals">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.7.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Basic scDNAseq analysis</h1>
            
      
      
      <div class="hidden name"><code>basic_scdnaseq.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">signals</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>Although this package was put together to infer allele or haplotype
specific copy number in scDNAseq data, it has a lot of methods for
visualization and analysis that should be useful for any type of
scDNAseq data, including when you only have total copy number calls.
Here we’ll go through the available plotting functions and various
utilities to summarize the data.</p>
</div>
<div class="section level2">
<h2 id="data-input">Data input<a class="anchor" aria-label="anchor" href="#data-input"></a>
</h2>
<p>The main requirement is a dataframe with the following columns:
<code>chr</code>, <code>start</code>, <code>end</code>,
<code>cell_id</code>, <code>state</code>, <code>copy</code>. state is
the inferred total copy number state. copy values are GC-correceted,
ploidy corrected normalized read counts that would be used to infer the
states. This may difer depending on the tool you use. We provide some
example data (<code>CNbins</code>) with this package. Most functions
will work with either this kind of dataframe or a <code>signals</code>
object, here we’ll just use the <code>CNbins</code> dataframe.</p>
</div>
<div class="section level2">
<h2 id="plotting">Plotting<a class="anchor" aria-label="anchor" href="#plotting"></a>
</h2>
<div class="section level3">
<h3 id="single-cell-profiles">Single cell profiles<a class="anchor" aria-label="anchor" href="#single-cell-profiles"></a>
</h3>
<p>The first thing we can do is plot single cell copy number
profiles.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCNprofile.html">plotCNprofile</a></span><span class="op">(</span><span class="va">CNbins</span>, cellid <span class="op">=</span> <span class="st">"SA921-A90554A-R03-C44"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_scdnaseq_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>These can tweaked in various ways, including modifying the y-axis,
only plotting certain chromosomes or changing the point size etc:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCNprofile.html">plotCNprofile</a></span><span class="op">(</span><span class="va">CNbins</span>, cellid <span class="op">=</span> <span class="st">"SA921-A90554A-R03-C44"</span>, y_axis_trans <span class="op">=</span> <span class="st">"squashy"</span>, maxCN <span class="op">=</span> <span class="fl">25</span>, chrfilt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"6"</span>, <span class="st">"7"</span>, <span class="st">"8"</span><span class="op">)</span>, pointsize <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_scdnaseq_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>We can also plot all the cells using the <code>plotHeatmap</code>
function. This can optionally take in a phylogenetic/clustering tree in
newick format. If you don’t have this you can use the
<code>umap_clustering</code> function to cluster your data using umap
and hdbscan.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustering</span> <span class="op">&lt;-</span> <span class="fu">umap_clustering</span><span class="op">(</span><span class="va">CNbins</span>, field <span class="op">=</span> <span class="st">"copy"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h3>
<p>Then we can use this output in the <code>plotHeatmap</code> function.
If you leave the <code>tree</code> and <code>clustering</code> options
empty then <code>plotHeatmap</code> will fall back to use
<code>umap_clustering</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>, tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>, clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_scdnaseq_files/figure-html/unnamed-chunk-5-1.png" width="1152"></p>
<p>The default is to plot the <code>state</code> column, we can change
this to plot the <code>copy</code> raw data column.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>, tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>, clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span>, show_clone_label <span class="op">=</span> <span class="cn">F</span>, plotcol <span class="op">=</span> <span class="st">"copy"</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_scdnaseq_files/figure-html/unnamed-chunk-6-1.png" width="1152"></p>
</div>
</div>
<div class="section level2">
<h2 id="utilities">Utilities<a class="anchor" aria-label="anchor" href="#utilities"></a>
</h2>
<p>Compute consensus copy number profiles.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CNbins_consensus</span> <span class="op">&lt;-</span> <span class="fu">consensuscopynumber</span><span class="op">(</span><span class="va">CNbins</span>, cl <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span> <span class="co">#cell_id becomes clone_id</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotCNprofile.html">plotCNprofile</a></span><span class="op">(</span><span class="va">CNbins_consensus</span><span class="op">)</span></span></code></pre></div>
<p><img src="basic_scdnaseq_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Create segmentation from binned data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">segments</span> <span class="op">&lt;-</span> <span class="fu">create_segments</span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span></code></pre></div>
<p>Compute average copy number per chromosome or chromosome arm.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CNbins_chr</span> <span class="op">&lt;-</span> <span class="fu">per_chr_cn</span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span>
<span></span>
<span><span class="va">CNbins_arm</span> <span class="op">&lt;-</span> <span class="fu">per_chrarm_cn</span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
