<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Beta-Binomial model for signals • signals</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Beta-Binomial model for signals">
<meta property="og:description" content="signals">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.8.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Beta-Binomial model for signals</h1>
            
      
      
      <div class="hidden name"><code>BetaBinom.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">signals</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>This vignette demonstrates how to use a beta-binomial model instead
of a binomial model for the HMM emission model.</p>
</div>
<div class="section level2">
<h2 id="inference">Inference<a class="anchor" aria-label="anchor" href="#inference"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"haplotypes"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CNbins"</span><span class="op">)</span></span>
<span><span class="va">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span><span class="op">(</span><span class="va">haplotypes</span>, <span class="va">CNbins</span><span class="op">)</span></span>
<span><span class="va">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span>, likelihood <span class="op">=</span> <span class="st">"binomial"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">hscn</span><span class="op">)</span></span>
<span><span class="co">#&gt; Haplotype specific copy number object </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; Number of cells: 250</span></span>
<span><span class="co">#&gt; Bin size: 0.5 Mb </span></span>
<span><span class="co">#&gt; Inferred LOH error rate: 0.018 </span></span>
<span><span class="co">#&gt; Emission model for HMM: binomial </span></span>
<span><span class="co">#&gt; Average distance from median to expected BAF = 0.0025 </span></span>
<span><span class="co">#&gt; Average ploidy = 4 </span></span>
<span><span class="co">#&gt; Average number of segments = 79</span></span>
<span></span>
<span><span class="va">hscn_bb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span><span class="op">(</span><span class="va">CNbins</span>, <span class="va">haplotypes</span>, likelihood <span class="op">=</span> <span class="st">"betabinomial"</span><span class="op">)</span></span>
<span><span class="co">#&gt; VGLM    linear loop  1 :  loglikelihood = -312957.3901</span></span>
<span><span class="co">#&gt; VGLM    linear loop  2 :  loglikelihood = -308994.9722</span></span>
<span><span class="co">#&gt; VGLM    linear loop  3 :  loglikelihood = -308442.2238</span></span>
<span><span class="co">#&gt; VGLM    linear loop  4 :  loglikelihood = -308416.2805</span></span>
<span><span class="co">#&gt; VGLM    linear loop  5 :  loglikelihood = -308415.8977</span></span>
<span><span class="co">#&gt; VGLM    linear loop  6 :  loglikelihood = -308415.8943</span></span>
<span><span class="co">#&gt; VGLM    linear loop  7 :  loglikelihood = -308415.8942</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">hscn_bb</span><span class="op">)</span></span>
<span><span class="co">#&gt; Haplotype specific copy number object </span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt; Number of cells: 250</span></span>
<span><span class="co">#&gt; Bin size: 0.5 Mb </span></span>
<span><span class="co">#&gt; Inferred LOH error rate: 0.018 </span></span>
<span><span class="co">#&gt; Emission model for HMM: betabinomial </span></span>
<span><span class="co">#&gt;   Inferred over dispersion: 0.0096</span></span>
<span><span class="co">#&gt;   Tarones Z score: 181.439</span></span>
<span><span class="co">#&gt; Average distance from median to expected BAF = 0.0026 </span></span>
<span><span class="co">#&gt; Average ploidy = 4 </span></span>
<span><span class="co">#&gt; Average number of segments = 79</span></span></code></pre></div>
<p>We see that the model inferred a modest degree of overdispersion in
the data, but that this was siginificantly different from 0 dispersion
(Tarones Z-score is high).</p>
<p>Below I’ll plot the heatmaps for the two output, which in this case
are very similar.</p>
<div class="section level3">
<h3 id="binomial-model">Binomial model<a class="anchor" aria-label="anchor" href="#binomial-model"></a>
</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">hscn</span>, plotcol <span class="op">=</span> <span class="st">"state_BAF"</span>, plottree <span class="op">=</span> <span class="cn">FALSE</span>, spacer_cols <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="BetaBinom_files/figure-html/unnamed-chunk-3-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="beta-binomial-model">Beta-Binomial model<a class="anchor" aria-label="anchor" href="#beta-binomial-model"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">hscn_bb</span>, plotcol <span class="op">=</span> <span class="st">"state_BAF"</span>, plottree <span class="op">=</span> <span class="cn">FALSE</span>, spacer_cols <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span></code></pre></div>
<p><img src="BetaBinom_files/figure-html/unnamed-chunk-4-1.png" width="960"></p>
</div>
<div class="section level3">
<h3 id="qc-plot">QC plot<a class="anchor" aria-label="anchor" href="#qc-plot"></a>
</h3>
<p>To interrogate the difference between the binomial and beta-binomial
models we can plot the BAF and overlay the two models.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotBBfit</span><span class="op">(</span><span class="va">hscn_bb</span><span class="op">)</span></span></code></pre></div>
<p><img src="BetaBinom_files/figure-html/unnamed-chunk-5-1.png" width="960"></p>
<p>We can check how similar the results are by comparing the two
dataframes. We find for this data that the results are very similar.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1093750      20</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">hscn_bb</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1093750      20</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="fu">orderdf</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span><span class="op">)</span>, <span class="fu">orderdf</span><span class="op">(</span><span class="va">hscn_bb</span><span class="op">$</span><span class="va">data</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Component \"alleleA\": 'is.NA' value mismatch: 17536 in current 17517 in target"    </span></span>
<span><span class="co">#&gt;  [2] "Component \"alleleB\": 'is.NA' value mismatch: 17536 in current 17517 in target"    </span></span>
<span><span class="co">#&gt;  [3] "Component \"totalcounts\": 'is.NA' value mismatch: 17536 in current 17517 in target"</span></span>
<span><span class="co">#&gt;  [4] "Component \"BAF\": 'is.NA' value mismatch: 17536 in current 17517 in target"        </span></span>
<span><span class="co">#&gt;  [5] "Component \"state_min\": Mean relative difference: 0.6055791"                       </span></span>
<span><span class="co">#&gt;  [6] "Component \"A\": Mean relative difference: 0.5750316"                               </span></span>
<span><span class="co">#&gt;  [7] "Component \"B\": Mean relative difference: 0.7174441"                               </span></span>
<span><span class="co">#&gt;  [8] "Component \"state_AS_phased\": 32657 string mismatches"                             </span></span>
<span><span class="co">#&gt;  [9] "Component \"state_AS\": 2802 string mismatches"                                     </span></span>
<span><span class="co">#&gt; [10] "Component \"LOH\": 1828 string mismatches"                                          </span></span>
<span><span class="co">#&gt; [11] "Component \"phase\": 29571 string mismatches"                                       </span></span>
<span><span class="co">#&gt; [12] "Component \"state_phase\": 30538 string mismatches"                                 </span></span>
<span><span class="co">#&gt; [13] "Component \"state_BAF\": Mean relative difference: 1.011266"</span></span></code></pre></div>
<p>Another thing we can check is the total number of segments identified
by the HMM. Using the beta-binomial model should reduce the influence of
any noisy regions and thus we might expect fewer segments. We do observe
fewer segments in the beta-binomial model.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">segs</span> <span class="op">&lt;-</span> <span class="fu">create_segments</span><span class="op">(</span><span class="va">hscn</span><span class="op">$</span><span class="va">data</span>, field <span class="op">=</span> <span class="st">"state_AS_phased"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">segs_bb</span> <span class="op">&lt;-</span> <span class="fu">create_segments</span><span class="op">(</span><span class="va">hscn_bb</span><span class="op">$</span><span class="va">data</span>, field <span class="op">=</span> <span class="st">"state_AS_phased"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">segs</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 39022     5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">segs_bb</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 35261     5</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>With this dataset, the difference between the nbinomial model and
beta-binomial model is minimal but produces a statistically superior
fit, in lower quality datasets with for example lower depth of coverage
the beta-binomial model may may provide more noticeable differences.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
