<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced Heatmap Visualization • signals</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced Heatmap Visualization">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">signals</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.14.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN-RNA.html">Allele Specific Copy Number Inference in scRNAseq</a>
    </li>
    <li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/basic_scdnaseq.html">Basic scDNAseq analysis</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for signals</a>
    </li>
    <li>
      <a href="../articles/heatmap_visualization.html">Advanced Heatmap Visualization</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Advanced Heatmap Visualization</h1>
            
      

      <div class="hidden name"><code>heatmap_visualization.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">signals</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <code>signals</code> package provides powerful heatmap
visualization capabilities for copy number data.</p>
<ul>
<li>
<strong>Gene annotations</strong>: Highlight specific genes of
interest at the top of the heatmap</li>
<li>
<strong>Chromosome ideogram</strong>: Display cytoband patterns and
centromeric regions</li>
<li>
<strong>Full genome visualization</strong>: Show all genomic bins
including centromeric regions</li>
</ul>
<p>These features work with any copy number data, not just
haplotype-specific or allele-specific calls. This makes them useful for
visualizing total copy number profiles from any single-cell DNA
sequencing technology.</p>
</div>
<div class="section level2">
<h2 id="setup-and-data">Setup and Data<a class="anchor" aria-label="anchor" href="#setup-and-data"></a>
</h2>
<p>We’ll use the included <code>CNbins</code> test data, which contains
copy number calls for 250 cells across all human chromosomes. First,
let’s load the data and create a clustering for organized heatmap
visualization.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span>
<span><span class="va">CNbins</span> <span class="op">&lt;-</span> <span class="va">CNbins</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">!=</span> <span class="st">"Y"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CNbins</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr   start     end reads     copy state               cell_id</span></span>
<span><span class="co">#&gt; 1   1 2000001 2500000   536 2.960868     2 SA921-A90554A-R03-C44</span></span>
<span><span class="co">#&gt; 2   1 3000001 3500000   499 2.603584     2 SA921-A90554A-R03-C44</span></span>
<span><span class="co">#&gt; 3   1 4000001 4500000   569 2.216388     2 SA921-A90554A-R03-C44</span></span>
<span><span class="co">#&gt; 4   1 4500001 5000000   524 2.039672     2 SA921-A90554A-R03-C44</span></span>
<span><span class="co">#&gt; 5   1 5000001 5500000   576 2.210792     2 SA921-A90554A-R03-C44</span></span>
<span><span class="co">#&gt; 6   1 5500001 6000000   516 2.028686     2 SA921-A90554A-R03-C44</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">clustering</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap_clustering.html">umap_clustering</a></span><span class="op">(</span><span class="va">CNbins</span>, field <span class="op">=</span> <span class="st">"copy"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="../reference/umap_clustering.html">umap_clustering()</a></code> function performs UMAP
dimensionality reduction followed by HDBSCAN clustering. This produces a
phylogenetic tree and cluster assignments that organize cells by
similarity.</p>
</div>
<div class="section level2">
<h2 id="basic-heatmap-baseline">Basic Heatmap (Baseline)<a class="anchor" aria-label="anchor" href="#basic-heatmap-baseline"></a>
</h2>
<p>Let’s start with a standard heatmap.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-4-1.png" class="r-plt" alt="" width="864"></p>
<p>This heatmap shows copy number states across all genomic bins
(columns) and all cells (rows). Cells are organized by the hierarchical
clustering tree (displayed on the left). The color scheme represents
copy number states, with blue indicating losses and red indicating
gains.</p>
</div>
<div class="section level2">
<h2 id="gene-annotations">Gene Annotations<a class="anchor" aria-label="anchor" href="#gene-annotations"></a>
</h2>
<p>The <code>gene_annotations</code> parameter highlights specific genes
of interest on the heatmap. This is useful for focusing on known
cancer-associated genes or other loci of biological significance.</p>
<div class="section level3">
<h3 id="single-gene-annotation">Single Gene Annotation<a class="anchor" aria-label="anchor" href="#single-gene-annotation"></a>
</h3>
<p>Let’s highlight the TP53 tumor suppressor gene:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="st">"TP53"</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-5-1.png" class="r-plt" alt="" width="864"></p>
<p>A line connecting the gene label to its genomic position appears
above the heatmap. This makes it easy to identify where specific genes
are located relative to the copy number changes.</p>
</div>
<div class="section level3">
<h3 id="multiple-gene-annotations">Multiple Gene Annotations<a class="anchor" aria-label="anchor" href="#multiple-gene-annotations"></a>
</h3>
<p>We can highlight multiple genes simultaneously:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"KRAS"</span>, <span class="st">"BRCA1"</span><span class="op">)</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="customizing-gene-labels">Customizing Gene Labels<a class="anchor" aria-label="anchor" href="#customizing-gene-labels"></a>
</h3>
<p>You can adjust the appearance of gene annotations:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"MYC"</span><span class="op">)</span>,</span>
<span>            gene_annotation_fontsize <span class="op">=</span> <span class="fl">12</span>,</span>
<span>            gene_link_height <span class="op">=</span> <span class="fl">8</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-7-1.png" class="r-plt" alt="" width="864"></p>
<p>Key parameters: - <code>gene_annotation_fontsize</code>: Controls the
size of gene labels (default uses <code>annofontsize</code>) -
<code>gene_link_height</code>: Height of connector lines in mm (default
is 5 mm)</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-annotations-with-frequency-plot">Gene Annotations with Frequency Plot<a class="anchor" aria-label="anchor" href="#gene-annotations-with-frequency-plot"></a>
</h2>
<p>Gene annotations can be combined with other annotations like the
frequency plot, which shows how many cells have copy number changes at
each genomic position:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"KRAS"</span><span class="op">)</span>,</span>
<span>            plotfrequency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            frequencycutoff <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-8-1.png" class="r-plt" alt="" width="864"></p>
<p>In this visualization: - Gene labels appear at the very top - Below
that is the frequency plot showing recurrence of copy number changes -
The heatmap shows individual cell profiles</p>
</div>
<div class="section level2">
<h2 id="chromosome-ideogram-visualization">Chromosome Ideogram Visualization<a class="anchor" aria-label="anchor" href="#chromosome-ideogram-visualization"></a>
</h2>
<p>The <code>plotideogram</code> parameter adds a visual representation
of chromosome structure at the bottom of the heatmap, including cytoband
patterns and centromeric regions.</p>
<p>⚠️ <strong>Important</strong>: The <code>plotideogram</code>
parameter requires <code>plotallbins = TRUE</code> to display all
genomic bins including centromeric regions.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            plotallbins <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            plotideogram <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-9-1.png" class="r-plt" alt="" width="864"></p>
<p>The ideogram shows: - <strong>Chromosome bands</strong>: Different
cytogenetic regions (displayed as colored bands) - <strong>Centromeric
regions</strong>: Light grey areas where coverage is typically low or
absent - <strong>Chromosome structure</strong>: Provides genomic context
for interpreting copy number changes</p>
</div>
<div class="section level2">
<h2 id="combining-all-features">Combining All Features<a class="anchor" aria-label="anchor" href="#combining-all-features"></a>
</h2>
<p>Here’s the complete visualization combining all new features:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"MYC"</span>, <span class="st">"KRAS"</span><span class="op">)</span>,</span>
<span>            plotallbins <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            plotideogram <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            plotfrequency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            frequencycutoff <span class="op">=</span> <span class="fl">2</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="864"></p>
<p>This comprehensive visualization shows: 1. <strong>Top</strong>: Gene
labels for key cancer-associated genes 2. <strong>Upper middle</strong>:
Frequency plot showing recurrent copy number changes 3.
<strong>Center</strong>: Individual cell copy number heatmap 4.
<strong>Bottom</strong>: Chromosome ideogram with cytoband and
centromere information</p>
<p>This arrangement provides multiple levels of information: individual
cell profiles, population-level recurrence patterns, genomic context,
and functional annotation (genes).</p>
</div>
<div class="section level2">
<h2 id="customization-options">Customization Options<a class="anchor" aria-label="anchor" href="#customization-options"></a>
</h2>
<div class="section level3">
<h3 id="genome-assembly">Genome Assembly<a class="anchor" aria-label="anchor" href="#genome-assembly"></a>
</h3>
<p>By default, the package uses hg19 (GRCh37) for gene positions and
cytoband information. You can switch to hg38 (GRCh38):</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="st">"TP53"</span>,</span>
<span>            genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="ideogram-height">Ideogram Height<a class="anchor" aria-label="anchor" href="#ideogram-height"></a>
</h3>
<p>Adjust the height of the ideogram annotation:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            plotallbins <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            plotideogram <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            ideogram_height <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-12-1.png" class="r-plt" alt="" width="864"></p>
<p>The <code>ideogram_height</code> parameter specifies height in cm
(default is 0.3 cm).</p>
</div>
<div class="section level3">
<h3 id="centromere-color">Centromere Color<a class="anchor" aria-label="anchor" href="#centromere-color"></a>
</h3>
<p>Customize the appearance of centromeric regions:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            plotallbins <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            plotideogram <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            centromere_col <span class="op">=</span> <span class="st">"#E0E0E0"</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-13-1.png" class="r-plt" alt="" width="864"></p>
</div>
</div>
<div class="section level2">
<h2 id="practical-use-cases">Practical Use Cases<a class="anchor" aria-label="anchor" href="#practical-use-cases"></a>
</h2>
<div class="section level3">
<h3 id="cancer-genomics">Cancer Genomics<a class="anchor" aria-label="anchor" href="#cancer-genomics"></a>
</h3>
<p>Highlight common tumor suppressors and oncogenes:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cancer_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"RB1"</span>, <span class="st">"PTEN"</span>, <span class="st">"BRCA1"</span>, <span class="st">"BRCA2"</span>, <span class="st">"MYC"</span>, <span class="st">"KRAS"</span>, <span class="st">"PIK3CA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="va">cancer_genes</span>,</span>
<span>            plotfrequency <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            frequencycutoff <span class="op">=</span> <span class="fl">3</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-14-1.png" class="r-plt" alt="" width="864"></p>
</div>
<div class="section level3">
<h3 id="focused-chromosomal-region">Focused Chromosomal Region<a class="anchor" aria-label="anchor" href="#focused-chromosomal-region"></a>
</h3>
<p>Display a specific chromosome with gene annotations:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Filter to chromosome 17</span></span>
<span><span class="va">CNbins_chr17</span> <span class="op">&lt;-</span> <span class="va">CNbins</span><span class="op">[</span><span class="va">CNbins</span><span class="op">$</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"17"</span>, <span class="op">]</span></span>
<span><span class="va">clustering_chr17</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/umap_clustering.html">umap_clustering</a></span><span class="op">(</span><span class="va">CNbins_chr17</span>, field <span class="op">=</span> <span class="st">"copy"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plotHeatmap.html">plotHeatmap</a></span><span class="op">(</span><span class="va">CNbins_chr17</span>,</span>
<span>            gene_annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TP53"</span>, <span class="st">"BRCA1"</span><span class="op">)</span>,</span>
<span>            tree <span class="op">=</span> <span class="va">clustering_chr17</span><span class="op">$</span><span class="va">tree</span>,</span>
<span>            clusters <span class="op">=</span> <span class="va">clustering_chr17</span><span class="op">$</span><span class="va">clustering</span><span class="op">)</span></span></code></pre></div>
<p><img src="heatmap_visualization_files/figure-html/unnamed-chunk-15-1.png" class="r-plt" alt="" width="864"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
